# Inputs and formats

```{r setup, include=FALSE}
WORDS_TO_IGNORE <- c("nd")
source("knitr-options.R")
source("spelling-check.R")
```

## In {bigstatsr}

The format provided in package {bigstatsr} is called a Filebacked Big Matrix (FBM). It is an on-disk matrix format which is accessed through memory-mapping.

```{r, echo=FALSE}
knitr::include_graphics("https://raw.githubusercontent.com/privefl/R-presentation/master/memory-solution.svg")
```

How memory-mapping works:

- when you access the 1st element (1st row, 1st col), it accesses a block (say the first column) from disk into memory (RAM)
- when you access the 2nd element (2nd row, 1st col), it is already in memory so it is accessed very fast
- when you access the second column, you access from disk again (once)
- you can access many columns like that, until you do not have enough memory anymore
- some space is freed automatically so that new columns can be accessed into memory
- everything is seamlessly managed by the operating system (OS)
- it is also very convenient for parallelism as data is shared between processes

All the elements of an FBM have the same type; supported types are:

 - `"double"` (the default, double precision -- 64 bits)
 - `"float"` (single precision -- 32 bits)
 - `"integer"` (signed, so between $\text{-}2^{31}$ and $2^{31} \text{ - } 1$)
 - `"unsigned short"`: can store integer values from $0$ to $65535$
 - `"raw"` or `"unsigned char"`: can store integer values from $0$ to $255$. It is the basis for class *FBM.code256* in order to access 256 arbitrary different numeric values. It is used in package {bigsnpr} (see below).


## In {bigsnpr}

Package {bigsnpr} uses a class called `bigSNP` for representing SNP data. A `bigSNP` object is merely a list with the following elements:

- `$genotypes`: A [`FBM.code256`](https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html). Rows are samples and columns are genetic variants. This stores genotype calls or *dosages* (rounded to 2 decimal places).
- `$fam`: A `data.frame` with some information on the samples.
- `$map`: A `data.frame` with some information on the genetic variants.

The code used in class *FBM.code256* for imputed data is e.g. 
```{r, R.options=list(max.print=1000)}
bigsnpr::CODE_DOSAGE
```
where the first four elements are used to store genotype calls, the next three to store imputed allele counts, and the next 201 values to store dosages rounded to 2 decimal places. This allows for handling many types of data while storing each elements using one byte only (x4 compared to bed files, but /8 compared to doubles). 

Since v1.0, package {bigsnpr} also provides functions for directly working on bed files with a small percentage of missing values [@prive2020efficient].

:::: {.infobox .info}
If there is a demand for it, I might extend functions in {bigsnpr} to handle more types of FBMs than only *FBM.code256*. We have started talking about this in [this issue](https://github.com/privefl/bigsnpr/issues/185).
::::

## Getting an FBM or bigSNP object {#getting-FBM}

- The easiest way to get an FBM is to use the constructor function `FBM()` or the converter `as_FBM()`.

- To read an FBM from a large text file, you can use function `big_read()` (see [this vignette](https://privefl.github.io/bigstatsr/articles/read-FBM-from-file.html)).

- To read a bigSNP object from *bed*/*bim*/*fam* files, you can use functions `snp_readBed()` and `snp_readBed2()` (the second can read a subset of individuals/variants and use parallelism).

- To read *BGEN* files, you can use function `snp_readBGEN()`. This function takes around 40 minutes to read 1M variants for 400K individuals using 15 cores. Note that this function works only for BGEN v1.2 with probabilities stored as 8 bits (cf. [this issue](https://github.com/privefl/bigsnpr/issues/140)), which is the case for e.g. the UK Biobank files.

- To read any format used in genetics, you can always convert blocks of the data to text files using PLINK, read these using `bigreadr::fread2()`, and fill part of the resulting FBM. For example, see [the code I used to convert the iPSYCH imputed data from the RICOPILI pipeline to my bigSNP format](https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/1-ricopili-to-bigsnp.R).
