# Inputs and formats

```{r setup, include=FALSE}
source("knitr-options.R")
source("spelling-check.R")
```

## In {bigstatsr}

The format provided in package {bigstatsr} is called a Filebacked Big Matrix (FBM). It is an on-disk matrix format which is accessed through memory-mapping.

:::: {.infobox .info}
Memory-mapping accesses the elements you need from disk to memory when needed. The second time you access the same elements, they are directly accessed from memory while there is enough memory for storing them. When no more memory is available, the OS frees some of the memory to make room for new elements you want to access.
Therefore, try to always perform as many operations on a subset of elements since data is accessed from disk only once. 
::::

All the elements of one FBM have the same type; supported types are:

 - `"double"` (the default, double precision -- 64 bits)
 
 - `"float"` (single precision -- 32 bits)
 
 - `"integer"` (signed, so between $-2^{31}$ and $2^{31} - 1$)
 
 - `"unsigned short"`: can store integer values from $0$ to $65535$.
 
 - `"raw"` or `"unsigned char"`: can store integer values from $0$ to $255$. It is the basis for class *FBM.code256* in order to access 256 arbitrary different numeric values. It is used in package {bigsnpr} (see below).
 
:::: {.infobox .info}
Standard matrix accessors are implemented for FBMs, so you can e.g. access the first column of `X` using `X[, 1]`. You can access the whole FBM as an R matrix in memory using `X[]`.
However, if the matrix is too large to fit in memory, you should always access only a subset of columns.
Note that the elements of the FBM are stored column-wise (as for a standard R matrix). Therefore, be careful not to access a subset of rows, since it will read non-contiguous elements from the whole matrix from disk.
::::


## In {bigsnpr}

This package uses a class called `bigSNP` for representing SNP data. A `bigSNP` object is merely a list with the following elements:

- `$genotypes`: A [`FBM.code256`](https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html). Rows are samples and columns are variants. This stores genotype calls or *dosages* (rounded to 2 decimal places).
- `$fam`: A `data.frame` with some information on the individuals.
- `$map`: A `data.frame` with some information on the variants.

:::: {.infobox .info}
The code used in class *FBM.code256* for imputed data is e.g. 
```{r}
bigsnpr::CODE_DOSAGE
```
where the first four elements are used to store genotype calls, the next three to store imputed allele counts, and the next 201 values to store dosages rounded to 2 decimal places. This allows for many types of data by storing each elements using only one byte (x4 compared to bed files, but /8 compared to doubles). 
::::

Since v1.0, package {bigsnpr} also provides functions for directly working on bed files with a few missing values [@prive2020efficient].

:::: {.infobox .info}
If there is a demand for it, I might extend functions in {bigsnpr} to handle more types of FBMs than only *FBM.code256*. We have started talking about this in [this issue](https://github.com/privefl/bigsnpr/issues/185).
::::

## Getting a FBM or bigSNP object

- The easiest way to get an FBM is to use the constructor function `FBM()` or the converter `as_FBM()`. 

- To read an FBM from a large text file, you can use function `big_read()` (see [this vignette](https://privefl.github.io/bigstatsr/articles/read-FBM-from-file.html)).

- To read a bigSNP object from *bed*/*bim*/*fam* files, you can use functions `snp_readBed()` and `snp_readBed2()` (the second one allows using only a subset of individuals/variants and using parallelism). 

- To read *BGEN* files, you can use function `snp_readBGEN()`. This function takes around 40 minutes to read 1M variants for 400K individuals using 15 cores. Note that works only for BGEN V1.2 with probabilities stored as 8 bits (which is the case of e.g. the UK Biobank files, see [this issue](https://github.com/privefl/bigsnpr/issues/140)).

- To read any format used in genetics, you can always convert blocks of the data to text files using PLINK, read these using `bigreadr::fread2()` and filling a part of the resulting FBM. [TODO: export the code to convert from RICOPILI imputed data]
