# Introduction

```{r setup, include=FALSE}
WORDS_TO_IGNORE <- c("pak", "EWAS", "methylation")
source("knitr-options.R")
source("spelling-check.R")
```

## Main motivation for developing {bigstatsr} and {bigsnpr}

The main motivation was for me to be able to run all my analyses within `r icon::fa_r_project()`.
I was frustrated by having to use all these different software, with different input formats, and requiring text files for parameters.
This made it hard for me to build a chain of analyses, to perform some exploratory analyses, or to use familiar `r icon::fa_r_project()` packages.
Also, I wanted to develop new methods, which seemed very hard to do without using a simple matrix-like format.
Thus I started developing `r icon::fa_r_project()` package {bigsnpr} at the beginning of my thesis.

At some point, I realized that many functions (to perform e.g. [GWAS](https://en.wikipedia.org/wiki/Genome-wide_association_study), [PCA](https://en.wikipedia.org/wiki/Principal_component_analysis), summary statistics) were not really specific to genotype data.
Indeed, a TWAS or an EWAS are not conceptually very different from a GWAS; one can also perform PCA on e.g. DNA methylation data.
Therefore I decided to move all these functions that could be used on any data stored as a matrix into a new `r icon::fa_r_project()` package, {bigstatsr}.
This is why there are two packages, where {bigstatsr} can basically be used by any field using data stored as large matrices, while {bigsnpr} provides some tools rather specific to genotype data, largely building on top of {bigstatsr}. The initial description of the two packages is available in @prive2017efficient.

:::: {.infobox .info}

Functions starting with `big_` are part of {bigstatsr}, while functions starting with `snp_` or `bed_` are part of {bigsnpr}.

::::

## Features

There are now many things implemented in the packages. You can find a comprehensive list of available functions on the package website [of {bigstatsr}](https://privefl.github.io/bigstatsr/reference/index.html) and [of {bigsnpr}](https://privefl.github.io/bigsnpr/reference/index.html).

The next table presents an overview of common genetic analyses that are already implemented in {bigstatsr} and {bigsnpr}. This listing is inspired from table 1 of @visscher201710 "The Role of GWAS SNP Arrays in Human Genetic Discoveries".

| Analysis                          | Available in {bigstatsr} and {bigsnpr} | Still missing               | Citations              |
|-----------------------------------|----------------------------------------|-----------------------------|----------------------------------|
| Polygenic risk scores             | - penalized regressions <br> - (stacked) C+T <br> - LDpred2 <br> - lassosum2        | multi-ancestry training          | @prive2019efficient <br> @prive2019making <br> @prive2020ldpred2 <br> @prive2021identifying  |
| Population structure              | - principal component analysis <br> - ancestry inference <br> - fixation index ($F_{ST}$) <br> - local adaptation |           | @prive2017efficient <br> @prive2020efficient <br> @prive2021high <br> @prive2021using <br> @prive2020performing |
| GWAS                              | linear and logistic                    | - mixed models <br> - rare variant association                 | @prive2017efficient |
| Genome-wide assessment of LD      | - sparse correlation matrix <br> - optimal LD splitting    |          | @prive2017efficient <br> @prive2021optimal |
| Estimation of SNP heritability    | - LD score regression <br> - LDpred2-auto        |     | @prive2020ldpred2 <br> @prive2022inferring |
| Estimation of polygenicity        | LDpred2-auto                           |      | @prive2022inferring |
| Estimation of genetic correlation |                                        | need to extend LDpred2-auto |                            |
| Fine-mapping                      | LDpred2-auto                           |      | @prive2022inferring |
| GWAS summary imputation           |                                        | need to extend LDpred2-auto |                            |
| Mendelian randomization           |                                        | completely missing          |                          |
| Miscellaneous                     | - integration with PLINK <br> - format conversion <br> - imputation of genotyped variants <br> - matrix operations <br> - summaries       |              | @prive2017efficient |


## Installation

Both packages are available on CRAN, so you can use `install.packages()`:

```{r, eval=FALSE}
install.packages("bigstatsr")
install.packages("bigsnpr")
```

To install the latest versions (from GitHub), you can use {remotes}:

```{r, eval=FALSE}
# install.packages("remotes")
remotes::install_github("privefl/bigstatsr")
remotes::install_github("privefl/bigsnpr")
```


## Correct spellings

<blockquote class="twitter-tweet" data-lang="en" align="center" width="50%"><p lang="en" dir="ltr">A friendly reminder:<br>The correct spelling is<br> - bigstatsr -- not bigstatr / BIGstatsR<br> - bigsnpr -- not BIGsnpR / bigSNPr<br> - pcadapt -- not PCAdapt<br> - LDpred -- not LDPred<br><br>Thank you<br>The kittens thank you too <a href="https://t.co/S8wyE4G6BG">pic.twitter.com/S8wyE4G6BG</a></p>&mdash; Florian Priv√© (remote postdoc) (@privefl) <a href="https://twitter.com/privefl/status/1329109034218237953?ref_src=twsrc%5Etfw">November 18, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 
