<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Inputs and formats | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.29 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Inputs and formats | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Inputs and formats | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />

<meta name="author" content="Florian Privé" />


<meta name="date" content="2022-11-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="working-with-an-fbm.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#example-code"><i class="fa fa-check"></i><b>1.3</b> Example code</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.5</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-FBM"><i class="fa fa-check"></i><b>2.3</b> Getting an FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html"><i class="fa fa-check"></i><b>3</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="3.1" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>3.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="3.2" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>3.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="3.3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>3.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="3.4" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>3.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>4</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>4.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="4.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>4.2</b> Imputation</a></li>
<li class="chapter" data-level="4.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>4.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>5</b> Population structure</a>
<ul>
<li class="chapter" data-level="5.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>5.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="5.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>5.2</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>6</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>6.1</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html"><i class="fa fa-check"></i><b>7</b> Polygenic scores (PGS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#example-ldpred2-and-lassosum2"><i class="fa fa-check"></i><b>7.1</b> Example: LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#preparing-the-data"><i class="fa fa-check"></i><b>7.1.1</b> Preparing the data</a></li>
<li class="chapter" data-level="7.1.2" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#ldpred2"><i class="fa fa-check"></i><b>7.1.2</b> LDpred2</a></li>
<li class="chapter" data-level="7.1.3" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>7.1.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ipsych-data.html"><a href="ipsych-data.html"><i class="fa fa-check"></i><b>8</b> iPSYCH data</a>
<ul>
<li class="chapter" data-level="8.1" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-genomedk"><i class="fa fa-check"></i><b>8.1</b> Data on GenomeDK</a></li>
<li class="chapter" data-level="8.2" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-statistics-denmark"><i class="fa fa-check"></i><b>8.2</b> Data on Statistics Denmark</a></li>
<li class="chapter" data-level="8.3" data-path="ipsych-data.html"><a href="ipsych-data.html#warnings-about-the-data"><i class="fa fa-check"></i><b>8.3</b> Warnings about the data</a></li>
<li class="chapter" data-level="8.4" data-path="ipsych-data.html"><a href="ipsych-data.html#other-data-available"><i class="fa fa-check"></i><b>8.4</b> Other data available</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="inputs-and-formats" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Inputs and formats<a href="inputs-and-formats.html#inputs-and-formats" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="in-bigstatsr" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> In {bigstatsr}<a href="inputs-and-formats.html#in-bigstatsr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The format provided in package {bigstatsr} is called a Filebacked Big Matrix (FBM). It is an on-disk matrix format which is accessed through memory-mapping.</p>
<p><img src="https://raw.githubusercontent.com/privefl/R-presentation/master/memory-solution.svg" width="70%" style="display: block; margin: auto;" /></p>
<p>How memory-mapping works:</p>
<ul>
<li>when you access the 1st element (1st row, 1st col), it accesses a block (say the first column) from disk into memory (RAM)</li>
<li>when you access the 2nd element (2nd row, 1st col), it is already in memory so it is accessed very fast</li>
<li>when you access the second column, you access from disk again (once)</li>
<li>you can access many columns like that, until you do not have enough memory anymore</li>
<li>some space is freed automatically so that new columns can be accessed into memory</li>
<li>everything is seamlessly managed by the operating system (OS)</li>
<li>it is also very convenient for parallelism as data is shared between processes</li>
</ul>
<p>All the elements of an FBM have the same type; supported types are:</p>
<ul>
<li><code>"double"</code> (the default, double precision – 8 bytes per element)</li>
<li><code>"float"</code> (single precision – 4 bytes)</li>
<li><code>"integer"</code> (signed, so between <span class="math inline">\(\text{-}2^{31}\)</span> and (<span class="math inline">\(2^{31} \text{ - } 1\)</span>) – 4 bytes)</li>
<li><code>"unsigned short"</code>: can store integer values from <span class="math inline">\(0\)</span> to <span class="math inline">\(65535\)</span> (2 bytes)</li>
<li><code>"raw"</code> or <code>"unsigned char"</code>: can store integer values from <span class="math inline">\(0\)</span> to <span class="math inline">\(255\)</span> (1 byte). It is the basis for class <em>FBM.code256</em> in order to access 256 arbitrary different numeric values. It is used in package {bigsnpr} (see below).</li>
</ul>
</div>
<div id="in-bigsnpr" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> In {bigsnpr}<a href="inputs-and-formats.html#in-bigsnpr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Package {bigsnpr} uses a class called <code>bigSNP</code> for representing SNP data. A <code>bigSNP</code> object is merely a list with the following elements:</p>
<ul>
<li><code>$genotypes</code>: A <a href="https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html"><code>FBM.code256</code></a>. Rows are samples and columns are genetic variants. This stores genotype calls or <em>dosages</em> (rounded to 2 decimal places).</li>
<li><code>$fam</code>: A <code>data.frame</code> with some information on the samples.</li>
<li><code>$map</code>: A <code>data.frame</code> with some information on the genetic variants.</li>
</ul>
<p>The code used in class <em>FBM.code256</em> for imputed data is e.g. </p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="inputs-and-formats.html#cb1-1" aria-hidden="true" tabindex="-1"></a>bigsnpr<span class="sc">::</span>CODE_DOSAGE</span></code></pre></div>
<pre><code>#&gt;   [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04 0.05 0.06
#&gt;  [15] 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18 0.19 0.20
#&gt;  [29] 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31 0.32 0.33 0.34
#&gt;  [43] 0.35 0.36 0.37 0.38 0.39 0.40 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48
#&gt;  [57] 0.49 0.50 0.51 0.52 0.53 0.54 0.55 0.56 0.57 0.58 0.59 0.60 0.61 0.62
#&gt;  [71] 0.63 0.64 0.65 0.66 0.67 0.68 0.69 0.70 0.71 0.72 0.73 0.74 0.75 0.76
#&gt;  [85] 0.77 0.78 0.79 0.80 0.81 0.82 0.83 0.84 0.85 0.86 0.87 0.88 0.89 0.90
#&gt;  [99] 0.91 0.92 0.93 0.94 0.95 0.96 0.97 0.98 0.99 1.00 1.01 1.02 1.03 1.04
#&gt; [113] 1.05 1.06 1.07 1.08 1.09 1.10 1.11 1.12 1.13 1.14 1.15 1.16 1.17 1.18
#&gt; [127] 1.19 1.20 1.21 1.22 1.23 1.24 1.25 1.26 1.27 1.28 1.29 1.30 1.31 1.32
#&gt; [141] 1.33 1.34 1.35 1.36 1.37 1.38 1.39 1.40 1.41 1.42 1.43 1.44 1.45 1.46
#&gt; [155] 1.47 1.48 1.49 1.50 1.51 1.52 1.53 1.54 1.55 1.56 1.57 1.58 1.59 1.60
#&gt; [169] 1.61 1.62 1.63 1.64 1.65 1.66 1.67 1.68 1.69 1.70 1.71 1.72 1.73 1.74
#&gt; [183] 1.75 1.76 1.77 1.78 1.79 1.80 1.81 1.82 1.83 1.84 1.85 1.86 1.87 1.88
#&gt; [197] 1.89 1.90 1.91 1.92 1.93 1.94 1.95 1.96 1.97 1.98 1.99 2.00   NA   NA
#&gt; [211]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [225]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [239]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [253]   NA   NA   NA   NA</code></pre>
<p>where the first four elements are used to store genotype calls, the next three to store imputed allele counts, and the next 201 values to store dosages rounded to 2 decimal places. This allows for handling many types of data while storing each elements using one byte only (x4 compared to bed files, but /8 compared to doubles).</p>
<p>Since v1.0, package {bigsnpr} also provides functions for directly working on bed files with a small percentage of missing values <span class="citation">(<a href="#ref-prive2020efficient" role="doc-biblioref">Privé, Luu, Blum, McGrath, &amp; Vilhjálmsson, 2020</a>)</span>.</p>
<div class="infobox info">
<p>If there is a demand for it, I might extend functions in {bigsnpr} to handle more types of FBMs than only <em>FBM.code256</em>. We have started talking about this in <a href="https://github.com/privefl/bigsnpr/issues/185">this issue</a>.</p>
</div>
</div>
<div id="getting-FBM" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Getting an FBM or bigSNP object<a href="inputs-and-formats.html#getting-FBM" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>The easiest way to get an FBM is to use the constructor function <code>FBM()</code> or the converter <code>as_FBM()</code>.</p></li>
<li><p>To read an FBM from a large text file, you can use function <code>big_read()</code> (see <a href="https://privefl.github.io/bigstatsr/articles/read-FBM-from-file.html">this vignette</a>).</p></li>
<li><p>To read a bigSNP object from <em>bed</em>/<em>bim</em>/<em>fam</em> files, you can use functions <code>snp_readBed()</code> and <code>snp_readBed2()</code> (the second can read a subset of individuals/variants and use parallelism).</p></li>
<li><p>To read dosages from <em>BGEN</em> files, you can use function <code>snp_readBGEN()</code>. This function takes around 40 minutes to read 1M variants for 400K individuals using 15 cores. Note that this function works only for BGEN v1.2 with probabilities stored as 8 bits (cf. <a href="https://github.com/privefl/bigsnpr/issues/141#issuecomment-691285671">this issue</a>), which is the case for e.g. the UK Biobank files.</p></li>
<li><p>To read any format used in genetics, you can always convert blocks of the data to text files using PLINK, read these using <code>bigreadr::fread2()</code>, and fill part of the resulting FBM. For example, see <a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/1-ricopili-to-bigsnp.R">the code I used to convert the iPSYCH imputed data from the RICOPILI pipeline to my bigSNP format</a>.</p></li>
</ul>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-prive2020efficient" class="csl-entry">
Privé, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449–4457.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="working-with-an-fbm.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
