<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 Introduction | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.29 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 Introduction | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Introduction | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />

<meta name="author" content="Florian Privé" />


<meta name="date" content="2022-11-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="inputs-and-formats.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#example-code"><i class="fa fa-check"></i><b>1.3</b> Example code</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.5</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-FBM"><i class="fa fa-check"></i><b>2.3</b> Getting an FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html"><i class="fa fa-check"></i><b>3</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="3.1" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>3.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="3.2" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>3.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="3.3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>3.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="3.4" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>3.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>4</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>4.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="4.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>4.2</b> Imputation</a></li>
<li class="chapter" data-level="4.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>4.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>5</b> Population structure</a>
<ul>
<li class="chapter" data-level="5.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>5.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="5.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>5.2</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>6</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>6.1</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html"><i class="fa fa-check"></i><b>7</b> Polygenic scores (PGS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#example-ldpred2-and-lassosum2"><i class="fa fa-check"></i><b>7.1</b> Example: LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#preparing-the-data"><i class="fa fa-check"></i><b>7.1.1</b> Preparing the data</a></li>
<li class="chapter" data-level="7.1.2" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#ldpred2"><i class="fa fa-check"></i><b>7.1.2</b> LDpred2</a></li>
<li class="chapter" data-level="7.1.3" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>7.1.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ipsych-data.html"><a href="ipsych-data.html"><i class="fa fa-check"></i><b>8</b> iPSYCH data</a>
<ul>
<li class="chapter" data-level="8.1" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-genomedk"><i class="fa fa-check"></i><b>8.1</b> Data on GenomeDK</a></li>
<li class="chapter" data-level="8.2" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-statistics-denmark"><i class="fa fa-check"></i><b>8.2</b> Data on Statistics Denmark</a></li>
<li class="chapter" data-level="8.3" data-path="ipsych-data.html"><a href="ipsych-data.html#warnings-about-the-data"><i class="fa fa-check"></i><b>8.3</b> Warnings about the data</a></li>
<li class="chapter" data-level="8.4" data-path="ipsych-data.html"><a href="ipsych-data.html#other-data-available"><i class="fa fa-check"></i><b>8.4</b> Other data available</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">Chapter 1</span> Introduction<a href="introduction.html#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="main-motivation-for-developing-bigstatsr-and-bigsnpr" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> Main motivation for developing {bigstatsr} and {bigsnpr}<a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The main motivation was for me to be able to run all my analyses within <svg viewBox="0 0 581 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"></path></svg>.
I was frustrated by having to use all these different software, with different input formats, and requiring text files for parameters.
This made it hard for me to build a chain of analyses, to perform some exploratory analyses, or to use familiar <svg viewBox="0 0 581 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"></path></svg> packages.
Also, I wanted to develop new methods, which seemed very hard to do without using a simple matrix-like format.
Thus I started developing <svg viewBox="0 0 581 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"></path></svg> package {bigsnpr} at the beginning of my thesis.</p>
<p>At some point, I realized that many functions (to perform e.g. <a href="https://en.wikipedia.org/wiki/Genome-wide_association_study">GWAS</a>, <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">PCA</a>, summary statistics) were not really specific to genotype data.
Indeed, a TWAS or an EWAS are not conceptually very different from a GWAS; one can also perform PCA on e.g. DNA methylation data.
Therefore I decided to move all these functions that could be used on any data stored as a matrix into a new <svg viewBox="0 0 581 512" style="height:1em;position:relative;display:inline-block;top:.1em;" xmlns="http://www.w3.org/2000/svg"> <path d="M581 226.6C581 119.1 450.9 32 290.5 32S0 119.1 0 226.6C0 322.4 103.3 402 239.4 418.1V480h99.1v-61.5c24.3-2.7 47.6-7.4 69.4-13.9L448 480h112l-67.4-113.7c54.5-35.4 88.4-84.9 88.4-139.7zm-466.8 14.5c0-73.5 98.9-133 220.8-133s211.9 40.7 211.9 133c0 50.1-26.5 85-70.3 106.4-2.4-1.6-4.7-2.9-6.4-3.7-10.2-5.2-27.8-10.5-27.8-10.5s86.6-6.4 86.6-92.7-90.6-87.9-90.6-87.9h-199V361c-74.1-21.5-125.2-67.1-125.2-119.9zm225.1 38.3v-55.6c57.8 0 87.8-6.8 87.8 27.3 0 36.5-38.2 28.3-87.8 28.3zm-.9 72.5H365c10.8 0 18.9 11.7 24 19.2-16.1 1.9-33 2.8-50.6 2.9v-22.1z"></path></svg> package, {bigstatsr}.
This is why there are two packages, where {bigstatsr} can basically be used by any field using data stored as large matrices, while {bigsnpr} provides some tools rather specific to genotype data, largely building on top of {bigstatsr}. The initial description of the two packages is available in <span class="citation">Privé, Aschard, Ziyatdinov, &amp; Blum (<a href="#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span>.</p>
<div class="infobox info">
<p>Functions starting with <code>big_</code> are part of {bigstatsr}, while functions starting with <code>snp_</code> or <code>bed_</code> are part of {bigsnpr}.</p>
</div>
</div>
<div id="features" class="section level2 hasAnchor" number="1.2">
<h2><span class="header-section-number">1.2</span> Features<a href="introduction.html#features" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are now many things implemented in the packages. You can find a comprehensive list of available functions on the package website <a href="https://privefl.github.io/bigstatsr/reference/index.html">of {bigstatsr}</a> and <a href="https://privefl.github.io/bigsnpr/reference/index.html">of {bigsnpr}</a>.</p>
<p>The next table presents an overview of common genetic analyses that are already implemented in {bigstatsr} and {bigsnpr}. This listing is inspired from table 1 of <span class="citation">Visscher et al. (<a href="#ref-visscher201710" role="doc-biblioref">2017</a>)</span>.</p>
<table>
<colgroup>
<col width="25%" />
<col width="28%" />
<col width="21%" />
<col width="24%" />
</colgroup>
<thead>
<tr class="header">
<th>Analysis</th>
<th>Available in {bigstatsr} and {bigsnpr}</th>
<th>Still missing</th>
<th>Citations</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Polygenic risk scores</td>
<td>- penalized regressions <br> - (stacked) C+T <br> - LDpred2 <br> - lassosum2</td>
<td>multi-ancestry training</td>
<td><span class="citation">Privé, Aschard, &amp; Blum (<a href="#ref-prive2019efficient" role="doc-biblioref">2019</a>)</span> <br> <span class="citation">Privé, Vilhjálmsson, Aschard, &amp; Blum (<a href="#ref-prive2019making" role="doc-biblioref">2019</a>)</span> <br> <span class="citation">Privé, Arbel, &amp; Vilhjálmsson (<a href="#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span> <br> <span class="citation">Privé, Arbel, Aschard, &amp; Vilhjálmsson (<a href="#ref-prive2021identifying" role="doc-biblioref">2022</a>)</span></td>
</tr>
<tr class="even">
<td>Population structure</td>
<td>- principal component analysis <br> - ancestry inference <br> - fixation index (<span class="math inline">\(F_{ST}\)</span>) <br> - local adaptation</td>
<td></td>
<td><span class="citation">Privé et al. (<a href="#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span> <br> <span class="citation">Privé, Luu, Blum, McGrath, &amp; Vilhjálmsson (<a href="#ref-prive2020efficient" role="doc-biblioref">2020</a>)</span> <br> <span class="citation">Privé, Aschard, et al. (<a href="#ref-prive2021high" role="doc-biblioref">2022</a>)</span> <br> <span class="citation">Privé (<a href="#ref-prive2021using" role="doc-biblioref">2022</a>)</span> <br> <span class="citation">Privé, Luu, Vilhjálmsson, &amp; Blum (<a href="#ref-prive2020performing" role="doc-biblioref">2020</a>)</span></td>
</tr>
<tr class="odd">
<td>GWAS</td>
<td>linear and logistic</td>
<td>- mixed models <br> - rare variant association</td>
<td><span class="citation">Privé et al. (<a href="#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span></td>
</tr>
<tr class="even">
<td>Genome-wide assessment of LD</td>
<td>- sparse correlation matrix <br> - optimal LD splitting</td>
<td></td>
<td><span class="citation">Privé et al. (<a href="#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span> <br> <span class="citation">Privé (<a href="#ref-prive2021optimal" role="doc-biblioref">2021</a>)</span></td>
</tr>
<tr class="odd">
<td>Estimation of SNP heritability</td>
<td>- LD score regression <br> - LDpred2-auto</td>
<td></td>
<td><span class="citation">Privé, Arbel, et al. (<a href="#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span> <br> <span class="citation">Privé, Albiñana, Pasaniuc, &amp; Vilhjálmsson (<a href="#ref-prive2022inferring" role="doc-biblioref">2022</a>)</span></td>
</tr>
<tr class="even">
<td>Estimation of polygenicity</td>
<td>LDpred2-auto</td>
<td></td>
<td><span class="citation">Privé, Albiñana, et al. (<a href="#ref-prive2022inferring" role="doc-biblioref">2022</a>)</span></td>
</tr>
<tr class="odd">
<td>Estimation of genetic correlation</td>
<td></td>
<td>need to extend LDpred2-auto</td>
<td></td>
</tr>
<tr class="even">
<td>Fine-mapping</td>
<td>LDpred2-auto</td>
<td></td>
<td><span class="citation">Privé, Albiñana, et al. (<a href="#ref-prive2022inferring" role="doc-biblioref">2022</a>)</span></td>
</tr>
<tr class="odd">
<td>GWAS summary imputation</td>
<td></td>
<td>need to extend LDpred2-auto</td>
<td></td>
</tr>
<tr class="even">
<td>Mendelian randomization</td>
<td></td>
<td>completely missing</td>
<td></td>
</tr>
<tr class="odd">
<td>Miscellaneous</td>
<td>- integration with PLINK <br> - format conversion <br> - imputation of genotyped variants <br> - matrix operations <br> - summaries</td>
<td></td>
<td><span class="citation">Privé et al. (<a href="#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span></td>
</tr>
</tbody>
</table>
</div>
<div id="example-code" class="section level2 hasAnchor" number="1.3">
<h2><span class="header-section-number">1.3</span> Example code<a href="introduction.html#example-code" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>When you want to use a function for the first time, check the documentation and the examples in there (usually they are very short).</p></li>
<li><p>There are also many (longer) tutorials available (usually one with each paper), which be linked from here or are available at the packages’ websites.</p></li>
<li><p>Some other examples are provided in this extended documentation (in the next chapters).</p></li>
<li><p>All the code used in all my papers is available on GitHub. It mostly consists of R scripts based on {bigsnpr}, {bigstatsr}, the tidyverse, and the futureverse <span class="citation">(<a href="#ref-bengtsson2020unifying" role="doc-biblioref">Bengtsson, 2021</a>; <a href="#ref-prive2017efficient" role="doc-biblioref">Privé et al., 2018</a>; <a href="#ref-wickham2019welcome" role="doc-biblioref">Wickham et al., 2019</a>)</span>.</p></li>
</ul>
</div>
<div id="installation" class="section level2 hasAnchor" number="1.4">
<h2><span class="header-section-number">1.4</span> Installation<a href="introduction.html#installation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Both packages are available on CRAN, so you can use <code>install.packages()</code>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="introduction.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;bigstatsr&quot;</span>)</span>
<span id="cb3-2"><a href="introduction.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;bigsnpr&quot;</span>)</span></code></pre></div>
<p>To install the latest versions (from GitHub), you can use {remotes}:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="introduction.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;remotes&quot;)</span></span>
<span id="cb4-2"><a href="introduction.html#cb4-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;privefl/bigstatsr&quot;</span>)</span>
<span id="cb4-3"><a href="introduction.html#cb4-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;privefl/bigsnpr&quot;</span>)</span></code></pre></div>
</div>
<div id="correct-spellings" class="section level2 hasAnchor" number="1.5">
<h2><span class="header-section-number">1.5</span> Correct spellings<a href="introduction.html#correct-spellings" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<blockquote class="twitter-tweet" data-lang="en" align="center" width="50%">
<p lang="en" dir="ltr">
A friendly reminder:<br>The correct spelling is<br> - bigstatsr – not bigstatr / BIGstatsR<br> - bigsnpr – not BIGsnpR / bigSNPr<br> - pcadapt – not PCAdapt<br> - LDpred – not LDPred<br><br>Thank you<br>The kittens thank you too <a href="https://t.co/S8wyE4G6BG">pic.twitter.com/S8wyE4G6BG</a>
</p>
— Florian Privé (<span class="citation">(<a href="#ref-privefl" role="doc-biblioref"><strong>privefl?</strong></a>)</span>) <a href="https://twitter.com/privefl/status/1329109034218237953?ref_src=twsrc%5Etfw">November 18, 2020</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bengtsson2020unifying" class="csl-entry">
Bengtsson, H. (2021). <a href="https://doi.org/10.32614/RJ-2021-048">A unifying framework for parallel and distributed processing in <span>R</span> using futures</a>. <em><span>The R Journal</span></em>, <em>13</em>, 273–291.
</div>
<div id="ref-prive2021optimal" class="csl-entry">
Privé, F. (2021). <a href="https://doi.org/10.1093/bioinformatics/btab519"><span class="nocase">Optimal linkage disequilibrium splitting</span></a>. <em>Bioinformatics</em>, <em>38</em>, 255–256.
</div>
<div id="ref-prive2021using" class="csl-entry">
Privé, F. (2022). <a href="https://doi.org/10.1093/bioinformatics/btac348"><span class="nocase">Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics</span></a>. <em>Bioinformatics</em>, <em>38</em>, 3477–3480.
</div>
<div id="ref-prive2022inferring" class="csl-entry">
Privé, F., Albiñana, C., Pasaniuc, B., &amp; Vilhjálmsson, B.J. (2022). Inferring disease architecture and predictive ability with LDpred2-auto. <em>bioRxiv</em>. Retrieved from <a href="https://doi.org/10.1101/2022.10.10.511629">https://doi.org/10.1101/2022.10.10.511629</a>
</div>
<div id="ref-prive2021identifying" class="csl-entry">
Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136">Identifying and correcting for misspecifications in <span>GWAS</span> summary statistics and polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>, <em>3</em>, 100136.
</div>
<div id="ref-prive2020ldpred2" class="csl-entry">
Privé, F., Arbel, J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa1029"><span class="nocase"><span>LDpred2</span>: better, faster, stronger</span></a>. <em>Bioinformatics</em>, <em>36</em>, 5424–5431.
</div>
<div id="ref-prive2019efficient" class="csl-entry">
Privé, F., Aschard, H., &amp; Blum, M.G. (2019). <a href="https://doi.org/10.1534/genetics.119.302019">Efficient implementation of penalized regression for genetic risk prediction</a>. <em>Genetics</em>, <em>212</em>, 65–74.
</div>
<div id="ref-prive2021high" class="csl-entry">
Privé, F., Aschard, H., Carmi, S., Folkersen, L., Hoggart, C., O’Reilly, P.F., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.ajhg.2021.11.008">Portability of 245 polygenic scores when derived from the <span>UK Biobank</span> and applied to 9 ancestry groups from the same cohort</a>. <em>The American Journal of Human Genetics</em>, <em>109</em>, 12–23.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
Privé, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781–2787.
</div>
<div id="ref-prive2020efficient" class="csl-entry">
Privé, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449–4457.
</div>
<div id="ref-prive2020performing" class="csl-entry">
Privé, F., Luu, K., Vilhjálmsson, B.J., &amp; Blum, M.G. (2020). <a href="https://doi.org/10.1093/molbev/msaa053">Performing highly efficient genome scans for local adaptation with <span>R</span> package pcadapt version 4</a>. <em>Molecular Biology and Evolution</em>, <em>37</em>, 2153–2154.
</div>
<div id="ref-prive2019making" class="csl-entry">
Privé, F., Vilhjálmsson, B.J., Aschard, H., &amp; Blum, M.G.B. (2019). <a href="https://doi.org/10.1016/j.ajhg.2019.11.001">Making the most of clumping and thresholding for polygenic scores</a>. <em>The American Journal of Human Genetics</em>, <em>105</em>, 1213–1221.
</div>
<div id="ref-visscher201710" class="csl-entry">
Visscher, P.M., Wray, N.R., Zhang, Q., Sklar, P., McCarthy, M.I., Brown, M.A., &amp; Yang, J. (2017). <a href="https://doi.org/10.1016/j.ajhg.2017.06.005">10 years of <span>GWAS</span> discovery: Biology, function, and translation</a>. <em>The American Journal of Human Genetics</em>, <em>101</em>, 5–22.
</div>
<div id="ref-wickham2019welcome" class="csl-entry">
Wickham, H., Averick, M., Bryan, J., Chang, W., McGowan, L.D., François, R., et al.others. (2019). <a href="https://doi.org/10.21105/joss.01686">Welcome to the tidyverse</a>. <em>Journal of Open Source Software</em>, <em>4</em>, 1686.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="inputs-and-formats.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
