<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 Introduction | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 Introduction | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://privefl.github.io/bigsnpr-extdoc/" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Introduction | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />

<meta name="author" content="Florian Privé" />


<meta name="date" content="2021-03-23" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="inputs-and-formats.html"/>
<script src="libs/header-attrs-2.7.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/font-awesome-5.3.1/css/fontawesome-all.min.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.4</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-a-fbm-or-bigsnp-object"><i class="fa fa-check"></i><b>2.3</b> Getting a FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#conversion-and-quality-control-of-plink-files"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Exercise: preprocessing</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.2</b> Exercise: population structure</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Exercise: GWAS</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<div id="main-motivation-for-developing-bigstatsr-and-bigsnpr" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Main motivation for developing {bigstatsr} and {bigsnpr}</h2>
<p>The main motivation was for me to be able to run all my analyses within <i class="fab  fa-r-project "></i>.
I was frustrated by having to use all these different software, with different input formats, and requiring text files for parameters.
This made it hard for me to build a chain of analyses, to perform some exploratory analyses, or to use familiar <i class="fab  fa-r-project "></i> packages.
Also, I wanted to develop new methods, which seemed very hard to do without using a simple matrix-like format.</p>
<p>Thus I started developing <i class="fab  fa-r-project "></i> package {bigsnpr} at the beginning of my thesis.
At some point, I realized that many functions (to perform e.g. <a href="https://en.wikipedia.org/wiki/Genome-wide_association_study">GWAS</a>, <a href="https://en.wikipedia.org/wiki/Principal_component_analysis">PCA</a>, summary statistics) were not really specific to genotype data.
Indeed, a TWAS or an EWAS are not conceptually very different from a GWAS; one can also perform PCA on e.g. DNA methylation data.
Therefore I decided to move all these functions that could be used on any data stored as a matrix and build a new <i class="fab  fa-r-project "></i> package, {bigstatsr}.
This is why there are two packages, where {bigstatsr} can basically be used by any field using matrices, while {bigsnpr} provides some tools rather specific to genotype data, largely building on top of {bigstatsr}. The initial description of the two packages is available in <span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span>.</p>
<div class="infobox info">
<p>To know which function belongs to which package: functions starting with <code>big_</code> belongs to {bigstatsr} while functions starting with <code>snp_</code> or <code>bed_</code> belongs to {bigsnpr}.</p>
</div>
</div>
<div id="features" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Features</h2>
<p>There are now many things implemented in the packages. You can find a comprehensive list of available functions on the package website <a href="https://privefl.github.io/bigstatsr/reference/index.html">of {bigstatsr}</a> and <a href="https://privefl.github.io/bigsnpr/reference/index.html">of {bigsnpr}</a>.</p>
<p>The next table presents an overview of common genetic analyses that are already implemented in {bigstatsr} and {bigsnpr}. This listing is inspired from table 1 of <span class="citation"><a href="references.html#ref-visscher201710" role="doc-biblioref">Visscher et al.</a> (<a href="references.html#ref-visscher201710" role="doc-biblioref">2017</a>)</span> “The Role of GWAS SNP Arrays in Human Genetic Discoveries.”</p>
<table>
<colgroup>
<col width="25%" />
<col width="28%" />
<col width="21%" />
<col width="24%" />
</colgroup>
<thead>
<tr class="header">
<th>Analysis</th>
<th>Available in {bigstatsr} and {bigsnpr}</th>
<th>Still missing</th>
<th>Citations</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>GWAS</td>
<td>linear and logistic</td>
<td>- mixed models <br> - rare variant association</td>
<td><span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span></td>
</tr>
<tr class="even">
<td>Genome-wide assessment of LD</td>
<td>- sparse correlation matrix <br> - optimal LD splitting</td>
<td></td>
<td><span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2021optimal" role="doc-biblioref">Privé</a> (<a href="references.html#ref-prive2021optimal" role="doc-biblioref">2021</a>)</span></td>
</tr>
<tr class="odd">
<td>Estimation of SNP heritability</td>
<td>LDpred2-auto &amp; LD score regression</td>
<td>need proper validation</td>
<td><span class="citation"><a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">Privé, Arbel, and Vilhjálmsson</a> (<a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span></td>
</tr>
<tr class="even">
<td>Estimation of polygenicity</td>
<td>LDpred2-auto</td>
<td>need proper validation</td>
<td><span class="citation"><a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">Privé, Arbel, and Vilhjálmsson</a> (<a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span></td>
</tr>
<tr class="odd">
<td>Estimation of genetic correlation</td>
<td></td>
<td>need to extend LDpred2-auto</td>
<td></td>
</tr>
<tr class="even">
<td>Polygenic risk scores</td>
<td>- penalized regressions <br> - (stacked) C+T <br> - LDpred2</td>
<td></td>
<td><span class="citation"><a href="references.html#ref-prive2019efficient" role="doc-biblioref">Privé, Aschard, and Blum</a> (<a href="references.html#ref-prive2019efficient" role="doc-biblioref">2019</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2019making" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2019making" role="doc-biblioref">2019</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">Privé, Arbel, and Vilhjálmsson</a> (<a href="references.html#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span></td>
</tr>
<tr class="odd">
<td>Mendelian randomization</td>
<td></td>
<td>completely missing</td>
<td></td>
</tr>
<tr class="even">
<td>Population structure</td>
<td>- principal component analysis <br> - ancestry inference <br> - fixation index (<span class="math inline">\(F_{ST}\)</span>) <br> - local adaptation</td>
<td></td>
<td><span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span> &amp; <span class="citation"><a href="references.html#ref-prive2020efficient" role="doc-biblioref">Privé, Luu, Blum, et al.</a> (<a href="references.html#ref-prive2020efficient" role="doc-biblioref">2020</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2021high" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2021high" role="doc-biblioref">2021</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2020efficient" role="doc-biblioref">Privé, Luu, Blum, et al.</a> (<a href="references.html#ref-prive2020efficient" role="doc-biblioref">2020</a>)</span> <br> <span class="citation"><a href="references.html#ref-prive2020performing" role="doc-biblioref">Privé, Luu, Vilhjálmsson, et al.</a> (<a href="references.html#ref-prive2020performing" role="doc-biblioref">2020</a>)</span></td>
</tr>
<tr class="odd">
<td>Fine-mapping</td>
<td></td>
<td>completely missing</td>
<td></td>
</tr>
<tr class="even">
<td>Miscellaneous</td>
<td>- integration with PLINK <br> - format conversion <br> - imputation of genotyped variants <br> - matrix operations <br> - summaries</td>
<td></td>
<td><span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">Privé et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span></td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Installation</h2>
<p>Both packages are available on CRAN, so you can use <code>install.packages()</code>:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="introduction.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;bigstatsr&quot;</span>)</span>
<span id="cb1-2"><a href="introduction.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;bigsnpr&quot;</span>)</span></code></pre></div>
<p>To install the latest versions (from GitHub), you can use {remotes}:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="introduction.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;remotes&quot;)</span></span>
<span id="cb2-2"><a href="introduction.html#cb2-2" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;privefl/bigstatsr&quot;</span>)</span>
<span id="cb2-3"><a href="introduction.html#cb2-3" aria-hidden="true" tabindex="-1"></a>remotes<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;privefl/bigsnpr&quot;</span>)</span></code></pre></div>
<p>As an alternative to {remotes}, you can also try <a href="https://github.com/r-lib/pak">{pak}</a>, which gives more information and solve other issues.</p>
</div>
<div id="correct-spellings" class="section level2" number="1.4">
<h2><span class="header-section-number">1.4</span> Correct spellings</h2>
<blockquote class="twitter-tweet" data-lang="en" align="center" width="50%">
<p lang="en" dir="ltr">
A friendly reminder:<br>The correct spelling is<br> - bigstatsr – not bigstatr / BIGstatsR<br> - bigsnpr – not BIGsnpR / bigSNPr<br> - pcadapt – not PCAdapt<br> - LDpred – not LDPred<br><br>Thank you<br>The kittens thank you too <a href="https://t.co/S8wyE4G6BG">pic.twitter.com/S8wyE4G6BG</a>
</p>
— Florian Privé (remote postdoc) (<span class="citation">(<a href="#ref-privefl" role="doc-biblioref"><strong>privefl?</strong></a>)</span>) <a href="https://twitter.com/privefl/status/1329109034218237953?ref_src=twsrc%5Etfw">November 18, 2020</a>
</blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="inputs-and-formats.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
