<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Polygenic scores (PGS) | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.29 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Polygenic scores (PGS) | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Polygenic scores (PGS) | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc//images/bigstatsr.png" />

<meta name="author" content="Florian PrivÃ©" />


<meta name="date" content="2022-11-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="genome-wide-association-study-gwas.html"/>
<link rel="next" href="ipsych-data.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#example-code"><i class="fa fa-check"></i><b>1.3</b> Example code</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.4</b> Installation</a></li>
<li class="chapter" data-level="1.5" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.5</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-FBM"><i class="fa fa-check"></i><b>2.3</b> Getting an FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html"><i class="fa fa-check"></i><b>3</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="3.1" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>3.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="3.2" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>3.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="3.3" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>3.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="3.4" data-path="working-with-an-fbm.html"><a href="working-with-an-fbm.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>3.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>4</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>4.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="4.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>4.2</b> Imputation</a></li>
<li class="chapter" data-level="4.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>4.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>5</b> Population structure</a>
<ul>
<li class="chapter" data-level="5.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>5.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="5.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>5.2</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>6</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>6.1</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html"><i class="fa fa-check"></i><b>7</b> Polygenic scores (PGS)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#example-ldpred2-and-lassosum2"><i class="fa fa-check"></i><b>7.1</b> Example: LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="7.1.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#preparing-the-data"><i class="fa fa-check"></i><b>7.1.1</b> Preparing the data</a></li>
<li class="chapter" data-level="7.1.2" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#ldpred2"><i class="fa fa-check"></i><b>7.1.2</b> LDpred2</a></li>
<li class="chapter" data-level="7.1.3" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>7.1.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="ipsych-data.html"><a href="ipsych-data.html"><i class="fa fa-check"></i><b>8</b> iPSYCH data</a>
<ul>
<li class="chapter" data-level="8.1" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-genomedk"><i class="fa fa-check"></i><b>8.1</b> Data on GenomeDK</a></li>
<li class="chapter" data-level="8.2" data-path="ipsych-data.html"><a href="ipsych-data.html#data-on-statistics-denmark"><i class="fa fa-check"></i><b>8.2</b> Data on Statistics Denmark</a></li>
<li class="chapter" data-level="8.3" data-path="ipsych-data.html"><a href="ipsych-data.html#warnings-about-the-data"><i class="fa fa-check"></i><b>8.3</b> Warnings about the data</a></li>
<li class="chapter" data-level="8.4" data-path="ipsych-data.html"><a href="ipsych-data.html#other-data-available"><i class="fa fa-check"></i><b>8.4</b> Other data available</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="polygenic-scores-pgs" class="section level1 hasAnchor" number="7">
<h1><span class="header-section-number">Chapter 7</span> Polygenic scores (PGS)<a href="polygenic-scores-pgs.html#polygenic-scores-pgs" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>PGS methods are the main topic of my research work. These are the main methods currently available in the packages:</p>
<ul>
<li><p>penalized regressions, with individual-level data (<span class="citation">PrivÃ©, Aschard, et al. (<a href="#ref-prive2019efficient" role="doc-biblioref">2019</a>)</span> + <a href="https://privefl.github.io/bigstatsr/articles/penalized-regressions.html">tutorial</a>)</p></li>
<li><p>Clumping and Thresholding (C+T) and Stacked C+T (SCT), with summary statistics and individual level data (<span class="citation">PrivÃ©, VilhjÃ¡lmsson, et al. (<a href="#ref-prive2019making" role="doc-biblioref">2019</a>)</span> + <a href="https://privefl.github.io/bigsnpr/articles/SCT.html">tutorial</a>)</p></li>
<li><p>LDpred2, with summary statistics (<span class="citation">PrivÃ©, Arbel, et al. (<a href="#ref-prive2020ldpred2" role="doc-biblioref">2020</a>)</span> + <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">tutorial</a>)</p></li>
<li><p>lassosum2, with the same input data as LDpred2 (<span class="citation">PrivÃ©, Arbel, et al. (<a href="#ref-prive2021identifying" role="doc-biblioref">2022</a>)</span> + <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html#lassosum2-grid-of-models">tutorial</a>)</p></li>
</ul>
<div class="infobox info">
<p>You can now use LDpred2-auto for inference (<span class="citation">PrivÃ©, AlbiÃ±ana, et al. (<a href="#ref-prive2022inferring" role="doc-biblioref">2022</a>)</span> + <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html#inference-with-ldpred2-auto">tutorial</a>).</p>
</div>
<div id="example-ldpred2-and-lassosum2" class="section level2 hasAnchor" number="7.1">
<h2><span class="header-section-number">7.1</span> Example: LDpred2 and lassosum2<a href="polygenic-scores-pgs.html#example-ldpred2-and-lassosum2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>You should also check <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">the other tutorial</a> mentioned before.</p>
<div id="preparing-the-data" class="section level3 hasAnchor" number="7.1.1">
<h3><span class="header-section-number">7.1.1</span> Preparing the data<a href="polygenic-scores-pgs.html#preparing-the-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Let us first read the data produced in <a href="preprocessing.html#exo-preprocessing">4.3</a>:</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="polygenic-scores-pgs.html#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="polygenic-scores-pgs.html#cb81-1" aria-hidden="true" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb81-2"><a href="polygenic-scores-pgs.html#cb81-2" aria-hidden="true" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb81-3"><a href="polygenic-scores-pgs.html#cb81-3" aria-hidden="true" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span>
<span id="cb81-4"><a href="polygenic-scores-pgs.html#cb81-4" aria-hidden="true" tabindex="-1"></a>map <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">transmute</span>(obj.bigsnp<span class="sc">$</span>map,</span>
<span id="cb81-5"><a href="polygenic-scores-pgs.html#cb81-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">chr =</span> chromosome, <span class="at">pos =</span> physical.pos,</span>
<span id="cb81-6"><a href="polygenic-scores-pgs.html#cb81-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">a0 =</span> allele2, <span class="at">a1 =</span> allele1)</span></code></pre></div>
<p>Download some GWAS summary statistics for CAD that I derived from the UK Biobank <span class="citation">(<a href="#ref-bycroft2018uk" role="doc-biblioref">Bycroft et al., 2018</a>)</span>, and prepare them in the format required by LDpred2:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="polygenic-scores-pgs.html#cb82-1" aria-hidden="true" tabindex="-1"></a>gz <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb82-2"><a href="polygenic-scores-pgs.html#cb82-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;https://figshare.com/ndownloader/files/38077323&quot;</span>,</span>
<span id="cb82-3"><a href="polygenic-scores-pgs.html#cb82-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;sumstats_CAD_tuto.csv.gz&quot;</span>)</span>
<span id="cb82-4"><a href="polygenic-scores-pgs.html#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="fu">readLines</span>(gz, <span class="at">n =</span> <span class="dv">3</span>)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;chr,pos,rsid,allele1,allele2,freq,info,beta,se&quot;                                                   
#&gt; [2] &quot;1,721290,rs12565286,C,G,0.035889027911808,0.941918079726998,0.0361758959140647,0.0290865883937757&quot;
#&gt; [3] &quot;1,752566,rs3094315,A,G,0.840799909379283,0.997586884856296,-0.0340838522604864,0.0144572980122262&quot;</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="polygenic-scores-pgs.html#cb84-1" aria-hidden="true" tabindex="-1"></a>sumstats <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb84-2"><a href="polygenic-scores-pgs.html#cb84-2" aria-hidden="true" tabindex="-1"></a>  gz,</span>
<span id="cb84-3"><a href="polygenic-scores-pgs.html#cb84-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">select    =</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;pos&quot;</span>, <span class="st">&quot;allele2&quot;</span>, <span class="st">&quot;allele1&quot;</span>, <span class="st">&quot;beta&quot;</span>, <span class="st">&quot;se&quot;</span>, <span class="st">&quot;freq&quot;</span>, <span class="st">&quot;info&quot;</span>),</span>
<span id="cb84-4"><a href="polygenic-scores-pgs.html#cb84-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;chr&quot;</span>, <span class="st">&quot;pos&quot;</span>, <span class="st">&quot;a0&quot;</span>, <span class="st">&quot;a1&quot;</span>, <span class="st">&quot;beta&quot;</span>, <span class="st">&quot;beta_se&quot;</span>, <span class="st">&quot;freq&quot;</span>, <span class="st">&quot;info&quot;</span>))</span>
<span id="cb84-5"><a href="polygenic-scores-pgs.html#cb84-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-6"><a href="polygenic-scores-pgs.html#cb84-6" aria-hidden="true" tabindex="-1"></a><span class="co"># GWAS effective sample size for binary traits (4 / (1 / n_case + 1 / n_control))</span></span>
<span id="cb84-7"><a href="polygenic-scores-pgs.html#cb84-7" aria-hidden="true" tabindex="-1"></a><span class="co"># For quantitative traits, just use the total sample size for `n_eff`.</span></span>
<span id="cb84-8"><a href="polygenic-scores-pgs.html#cb84-8" aria-hidden="true" tabindex="-1"></a>sumstats<span class="sc">$</span>n_eff <span class="ot">&lt;-</span> <span class="dv">4</span> <span class="sc">/</span> (<span class="dv">1</span> <span class="sc">/</span> <span class="dv">20791</span> <span class="sc">+</span> <span class="dv">1</span> <span class="sc">/</span> <span class="dv">323124</span>) </span></code></pre></div>
<p>Note that we recommend to use imputed HapMap3+ variants when available, for which you can download some precomputed LD reference for European individuals based on the UK Biobank.
Otherwise use the genotyped variants as I am doing here.
Try to use an LD reference with at least 2000 individuals (I have only 1401 in this example).
Please see <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">this other tutorial</a> for more information.</p>
<p>Let us now match the variants in the GWAS summary statistics with the internal data we have:</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="polygenic-scores-pgs.html#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb85-2"><a href="polygenic-scores-pgs.html#cb85-2" aria-hidden="true" tabindex="-1"></a>info_snp <span class="ot">&lt;-</span> <span class="fu">snp_match</span>(sumstats, map, <span class="at">return_flip_and_rev =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb85-3"><a href="polygenic-scores-pgs.html#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">freq =</span> <span class="fu">ifelse</span>(<span class="st">`</span><span class="at">_REV_</span><span class="st">`</span>, <span class="dv">1</span> <span class="sc">-</span> freq, freq), </span>
<span id="cb85-4"><a href="polygenic-scores-pgs.html#cb85-4" aria-hidden="true" tabindex="-1"></a>         <span class="st">`</span><span class="at">_REV_</span><span class="st">`</span> <span class="ot">=</span> <span class="cn">NULL</span>, <span class="st">`</span><span class="at">_FLIP_</span><span class="st">`</span><span class="ot">=</span> <span class="cn">NULL</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb85-5"><a href="polygenic-scores-pgs.html#cb85-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>#&gt;   chr    pos a0 a1         beta    beta_se       freq      info    n_eff
#&gt; 1   1 752566  T  C  0.034083852 0.01445730 0.15920009 0.9975869 78136.41
#&gt; 2   1 785989  G  A  0.018289010 0.01549153 0.13007399 0.9913233 78136.41
#&gt; 3   1 798959  G  A  0.003331013 0.01307707 0.20524280 0.9734898 78136.41
#&gt; 4   1 947034  T  C -0.021202725 0.02838148 0.03595249 0.9924989 78136.41
#&gt;   _NUM_ID_.ss _NUM_ID_
#&gt; 1           2        2
#&gt; 2           4        4
#&gt; 3           5        5
#&gt; 4           6        6
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 340206 rows ]</code></pre>
<p>Check the summary statistics; some quality control may be needed:</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="polygenic-scores-pgs.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(info_snp<span class="sc">$</span>n_eff)    <span class="co"># all the same values, otherwise filter at 70% of max</span></span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-4-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="polygenic-scores-pgs.html#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(info_snp<span class="sc">$</span>info)     <span class="co"># very good imputation; filter e.g. at 0.7</span></span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-4-2.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="polygenic-scores-pgs.html#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(info_snp<span class="sc">$</span>freq)</span></code></pre></div>
<pre><code>#&gt;      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
#&gt; 0.0000027 0.1100250 0.2261357 0.2365588 0.3580946 0.9998703</code></pre>
<p>Then we can perform some quality control on the summary statistics by checking whether standard deviations inferred from the external GWAS summary statistics are consistent with the ones in the internal data we have:</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="polygenic-scores-pgs.html#cb91-1" aria-hidden="true" tabindex="-1"></a>af_ref <span class="ot">&lt;-</span> <span class="fu">big_colstats</span>(G, <span class="at">ind.col =</span> info_snp<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>, <span class="at">ncores =</span> NCORES)<span class="sc">$</span>sum <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> <span class="fu">nrow</span>(G))</span>
<span id="cb91-2"><a href="polygenic-scores-pgs.html#cb91-2" aria-hidden="true" tabindex="-1"></a>sd_ref <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="dv">2</span> <span class="sc">*</span> af_ref <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> af_ref))</span>
<span id="cb91-3"><a href="polygenic-scores-pgs.html#cb91-3" aria-hidden="true" tabindex="-1"></a>sd_ss <span class="ot">&lt;-</span> <span class="fu">with</span>(info_snp, <span class="dv">2</span> <span class="sc">/</span> <span class="fu">sqrt</span>(n_eff <span class="sc">*</span> beta_se<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> beta<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb91-4"><a href="polygenic-scores-pgs.html#cb91-4" aria-hidden="true" tabindex="-1"></a>is_bad <span class="ot">&lt;-</span></span>
<span id="cb91-5"><a href="polygenic-scores-pgs.html#cb91-5" aria-hidden="true" tabindex="-1"></a>  sd_ss <span class="sc">&lt;</span> (<span class="fl">0.5</span> <span class="sc">*</span> sd_ref) <span class="sc">|</span> sd_ss <span class="sc">&gt;</span> (sd_ref <span class="sc">+</span> <span class="fl">0.1</span>) <span class="sc">|</span> </span>
<span id="cb91-6"><a href="polygenic-scores-pgs.html#cb91-6" aria-hidden="true" tabindex="-1"></a>  sd_ss <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> sd_ref <span class="sc">&lt;</span> <span class="fl">0.05</span>  <span class="co"># basically filtering small MAF</span></span>
<span id="cb91-7"><a href="polygenic-scores-pgs.html#cb91-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-8"><a href="polygenic-scores-pgs.html#cb91-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb91-9"><a href="polygenic-scores-pgs.html#cb91-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">slice_sample</span>(<span class="fu">data.frame</span>(sd_ref, sd_ss, is_bad), <span class="at">n =</span> <span class="fl">50e3</span>)) <span class="sc">+</span></span>
<span id="cb91-10"><a href="polygenic-scores-pgs.html#cb91-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(sd_ref, sd_ss, <span class="at">color =</span> is_bad), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb91-11"><a href="polygenic-scores-pgs.html#cb91-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>(<span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb91-12"><a href="polygenic-scores-pgs.html#cb91-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_d</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb91-13"><a href="polygenic-scores-pgs.html#cb91-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb91-14"><a href="polygenic-scores-pgs.html#cb91-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Standard deviations in the reference set&quot;</span>,</span>
<span id="cb91-15"><a href="polygenic-scores-pgs.html#cb91-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Standard deviations derived from the summary statistics&quot;</span>,</span>
<span id="cb91-16"><a href="polygenic-scores-pgs.html#cb91-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">&quot;To remove?&quot;</span>)</span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-5-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>When using quantitative traits (linear regression instead of logistic regression for the GWAS), you need to replace <code>2</code> by <code>sd(y)</code> when computing <code>sd_ss</code> (equations 1 and 2 of <span class="citation">PrivÃ©, Arbel, et al. (<a href="#ref-prive2021identifying" role="doc-biblioref">2022</a>)</span>).</p>
<p>When allele frequencies are available in the GWAS summary statistics, you can use them (along with INFO scores) to get an even better match:</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="polygenic-scores-pgs.html#cb92-1" aria-hidden="true" tabindex="-1"></a>sd_af <span class="ot">&lt;-</span> <span class="fu">with</span>(info_snp, <span class="fu">sqrt</span>(<span class="dv">2</span> <span class="sc">*</span> freq <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> freq) <span class="sc">*</span> info))</span>
<span id="cb92-2"><a href="polygenic-scores-pgs.html#cb92-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-3"><a href="polygenic-scores-pgs.html#cb92-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">slice_sample</span>(<span class="fu">data.frame</span>(sd_af, sd_ss), <span class="at">n =</span> <span class="fl">50e3</span>)) <span class="sc">+</span></span>
<span id="cb92-4"><a href="polygenic-scores-pgs.html#cb92-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(sd_af, sd_ss), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb92-5"><a href="polygenic-scores-pgs.html#cb92-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>(<span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb92-6"><a href="polygenic-scores-pgs.html#cb92-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb92-7"><a href="polygenic-scores-pgs.html#cb92-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Standard deviations derived from allele frequencies&quot;</span>,</span>
<span id="cb92-8"><a href="polygenic-scores-pgs.html#cb92-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Standard deviations derived from the summary statistics&quot;</span>)</span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-6-1.png" width="70%" style="display: block; margin: auto;" />
You can still use the reference panel to do some quality control by comparing allele frequencies:</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="polygenic-scores-pgs.html#cb93-1" aria-hidden="true" tabindex="-1"></a>diff <span class="ot">&lt;-</span> af_ref <span class="sc">-</span> info_snp<span class="sc">$</span>freq</span>
<span id="cb93-2"><a href="polygenic-scores-pgs.html#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(diff, <span class="st">&quot;FD&quot;</span>, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.1</span>, <span class="fl">0.1</span>))</span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-7-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Then you can filter</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="polygenic-scores-pgs.html#cb94-1" aria-hidden="true" tabindex="-1"></a>is_bad2 <span class="ot">&lt;-</span></span>
<span id="cb94-2"><a href="polygenic-scores-pgs.html#cb94-2" aria-hidden="true" tabindex="-1"></a>  sd_ss <span class="sc">&lt;</span> (<span class="fl">0.7</span> <span class="sc">*</span> sd_af) <span class="sc">|</span> sd_ss <span class="sc">&gt;</span> (sd_af <span class="sc">+</span> <span class="fl">0.1</span>) <span class="sc">|</span> </span>
<span id="cb94-3"><a href="polygenic-scores-pgs.html#cb94-3" aria-hidden="true" tabindex="-1"></a>  sd_ss <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span> sd_af <span class="sc">&lt;</span> <span class="fl">0.05</span> <span class="sc">|</span></span>
<span id="cb94-4"><a href="polygenic-scores-pgs.html#cb94-4" aria-hidden="true" tabindex="-1"></a>  info_snp<span class="sc">$</span>info <span class="sc">&lt;</span> <span class="fl">0.7</span> <span class="sc">|</span> <span class="fu">abs</span>(diff) <span class="sc">&gt;</span> <span class="fl">0.07</span></span>
<span id="cb94-5"><a href="polygenic-scores-pgs.html#cb94-5" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(is_bad2)</span></code></pre></div>
<pre><code>#&gt; [1] 0.002410276</code></pre>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="polygenic-scores-pgs.html#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(is_bad, is_bad2)</span></code></pre></div>
<pre><code>#&gt;        is_bad2
#&gt; is_bad   FALSE   TRUE
#&gt;   FALSE 339138    410
#&gt;   TRUE     252    410</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="polygenic-scores-pgs.html#cb98-1" aria-hidden="true" tabindex="-1"></a>df_beta <span class="ot">&lt;-</span> info_snp[<span class="sc">!</span>is_bad2, ]</span></code></pre></div>
<p>Then, we compute the correlation for each chromosome (note that we are using only 4 chromosomes for faster running of this tutorial):</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="polygenic-scores-pgs.html#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {  <span class="co"># REPLACE BY 1:22</span></span>
<span id="cb99-2"><a href="polygenic-scores-pgs.html#cb99-2" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb99-3"><a href="polygenic-scores-pgs.html#cb99-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(chr)</span>
<span id="cb99-4"><a href="polygenic-scores-pgs.html#cb99-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb99-5"><a href="polygenic-scores-pgs.html#cb99-5" aria-hidden="true" tabindex="-1"></a>  corr0 <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>({</span>
<span id="cb99-6"><a href="polygenic-scores-pgs.html#cb99-6" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb99-7"><a href="polygenic-scores-pgs.html#cb99-7" aria-hidden="true" tabindex="-1"></a>    <span class="do">## indices in &#39;sumstats&#39;</span></span>
<span id="cb99-8"><a href="polygenic-scores-pgs.html#cb99-8" aria-hidden="true" tabindex="-1"></a>    ind.chr <span class="ot">&lt;-</span> <span class="fu">which</span>(df_beta<span class="sc">$</span>chr <span class="sc">==</span> chr)</span>
<span id="cb99-9"><a href="polygenic-scores-pgs.html#cb99-9" aria-hidden="true" tabindex="-1"></a>    <span class="do">## indices in &#39;G&#39;</span></span>
<span id="cb99-10"><a href="polygenic-scores-pgs.html#cb99-10" aria-hidden="true" tabindex="-1"></a>    ind.chr2 <span class="ot">&lt;-</span> df_beta<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>[ind.chr]</span>
<span id="cb99-11"><a href="polygenic-scores-pgs.html#cb99-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb99-12"><a href="polygenic-scores-pgs.html#cb99-12" aria-hidden="true" tabindex="-1"></a>    POS2 <span class="ot">&lt;-</span> <span class="fu">snp_asGeneticPos</span>(map<span class="sc">$</span>chr[ind.chr2], map<span class="sc">$</span>pos[ind.chr2], <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb99-13"><a href="polygenic-scores-pgs.html#cb99-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">snp_cor</span>(G, <span class="at">ind.col =</span> ind.chr2, <span class="at">size =</span> <span class="dv">3</span> <span class="sc">/</span> <span class="dv">1000</span>, <span class="at">infos.pos =</span> POS2, </span>
<span id="cb99-14"><a href="polygenic-scores-pgs.html#cb99-14" aria-hidden="true" tabindex="-1"></a>            <span class="at">ncores =</span> NCORES)</span>
<span id="cb99-15"><a href="polygenic-scores-pgs.html#cb99-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb99-16"><a href="polygenic-scores-pgs.html#cb99-16" aria-hidden="true" tabindex="-1"></a>  }, <span class="at">file =</span> <span class="fu">paste0</span>(<span class="st">&quot;tmp-data/corr_chr&quot;</span>, chr, <span class="st">&quot;.rds&quot;</span>))</span>
<span id="cb99-17"><a href="polygenic-scores-pgs.html#cb99-17" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>#&gt; [1] 1
#&gt;    user  system elapsed 
#&gt;   36.72    0.42   13.09 
#&gt; [1] 2
#&gt;    user  system elapsed 
#&gt;   48.61    0.29   14.96 
#&gt; [1] 3
#&gt;    user  system elapsed 
#&gt;   45.52    0.17   13.65 
#&gt; [1] 4
#&gt;    user  system elapsed 
#&gt;   38.14    0.05   11.25</code></pre>
<p>Then we create the on-disk sparse genome-wide correlation matrix (again using only the first 4 chromosomes, for speed in this tutorial; replace by <code>1:22</code>):</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="polygenic-scores-pgs.html#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {  <span class="co"># REPLACE BY 1:22</span></span>
<span id="cb101-2"><a href="polygenic-scores-pgs.html#cb101-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-3"><a href="polygenic-scores-pgs.html#cb101-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(chr)</span>
<span id="cb101-4"><a href="polygenic-scores-pgs.html#cb101-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-5"><a href="polygenic-scores-pgs.html#cb101-5" aria-hidden="true" tabindex="-1"></a>  corr0 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">&quot;tmp-data/corr_chr&quot;</span>, chr, <span class="st">&quot;.rds&quot;</span>))</span>
<span id="cb101-6"><a href="polygenic-scores-pgs.html#cb101-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb101-7"><a href="polygenic-scores-pgs.html#cb101-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (chr <span class="sc">==</span> <span class="dv">1</span>) {</span>
<span id="cb101-8"><a href="polygenic-scores-pgs.html#cb101-8" aria-hidden="true" tabindex="-1"></a>    ld <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(corr0<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb101-9"><a href="polygenic-scores-pgs.html#cb101-9" aria-hidden="true" tabindex="-1"></a>    corr <span class="ot">&lt;-</span> <span class="fu">as_SFBM</span>(corr0, <span class="st">&quot;tmp-data/corr&quot;</span>, <span class="at">compact =</span> <span class="cn">TRUE</span>)</span>
<span id="cb101-10"><a href="polygenic-scores-pgs.html#cb101-10" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb101-11"><a href="polygenic-scores-pgs.html#cb101-11" aria-hidden="true" tabindex="-1"></a>    ld <span class="ot">&lt;-</span> <span class="fu">c</span>(ld, Matrix<span class="sc">::</span><span class="fu">colSums</span>(corr0<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb101-12"><a href="polygenic-scores-pgs.html#cb101-12" aria-hidden="true" tabindex="-1"></a>    corr<span class="sc">$</span><span class="fu">add_columns</span>(corr0, <span class="fu">nrow</span>(corr))</span>
<span id="cb101-13"><a href="polygenic-scores-pgs.html#cb101-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb101-14"><a href="polygenic-scores-pgs.html#cb101-14" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>#&gt; [1] 1
#&gt; [1] 2
#&gt; [1] 3
#&gt; [1] 4</code></pre>
<p>To use the âcompactâ format for SFBMs, you need <code>packageVersion("bigsparser") &gt;= package_version("0.5")</code>.
Make sure to reinstall {bigsnpr} when updating {bigsparser} to this new version (to avoid crashes).</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="polygenic-scores-pgs.html#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">file.size</span>(corr<span class="sc">$</span>sbk) <span class="sc">/</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">3</span>  <span class="co"># file size in GB</span></span></code></pre></div>
<pre><code>#&gt; [1] 0.5756225</code></pre>
<p>Note that you will need at least the same memory as this file size (to keep it cached for faster processing) + some other memory for all the results returned. If you do not have enough memory, processing will be very slow (because you would read the data from disk all the time). If using HapMap3 variants, requesting 60 GB should be enough. For this small example, 8 GB of RAM should be enough.</p>
</div>
<div id="ldpred2" class="section level3 hasAnchor" number="7.1.2">
<h3><span class="header-section-number">7.1.2</span> LDpred2<a href="polygenic-scores-pgs.html#ldpred2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We can now run LD score regression:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="polygenic-scores-pgs.html#cb105-1" aria-hidden="true" tabindex="-1"></a>df_beta <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(df_beta, chr <span class="sc">%in%</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>)  <span class="co"># TO REMOVE (for speed here)</span></span>
<span id="cb105-2"><a href="polygenic-scores-pgs.html#cb105-2" aria-hidden="true" tabindex="-1"></a>(ldsc <span class="ot">&lt;-</span> <span class="fu">with</span>(df_beta, <span class="fu">snp_ldsc</span>(ld, <span class="fu">length</span>(ld), <span class="at">chi2 =</span> (beta <span class="sc">/</span> beta_se)<span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb105-3"><a href="polygenic-scores-pgs.html#cb105-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">sample_size =</span> n_eff, <span class="at">blocks =</span> <span class="cn">NULL</span>)))</span></code></pre></div>
<pre><code>#&gt;       int        h2 
#&gt; 0.9795999 0.0528327</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="polygenic-scores-pgs.html#cb107-1" aria-hidden="true" tabindex="-1"></a>ldsc_h2_est <span class="ot">&lt;-</span> ldsc[[<span class="st">&quot;h2&quot;</span>]]</span></code></pre></div>
<p>We can now run LDpred2-inf very easily:</p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="polygenic-scores-pgs.html#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LDpred2-inf</span></span>
<span id="cb108-2"><a href="polygenic-scores-pgs.html#cb108-2" aria-hidden="true" tabindex="-1"></a>beta_inf <span class="ot">&lt;-</span> <span class="fu">snp_ldpred2_inf</span>(corr, df_beta, ldsc_h2_est)</span>
<span id="cb108-3"><a href="polygenic-scores-pgs.html#cb108-3" aria-hidden="true" tabindex="-1"></a>pred_inf <span class="ot">&lt;-</span> <span class="fu">big_prodVec</span>(G, beta_inf, <span class="at">ind.col =</span> df_beta<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>)</span>
<span id="cb108-4"><a href="polygenic-scores-pgs.html#cb108-4" aria-hidden="true" tabindex="-1"></a><span class="fu">AUCBoot</span>(pred_inf, obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>CAD)</span></code></pre></div>
<pre><code>#&gt;      Mean      2.5%     97.5%        Sd 
#&gt; 0.5511041 0.5186029 0.5826846 0.0163816</code></pre>
<p>For LDpred2(-grid), this is the grid we recommend to use:</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="polygenic-scores-pgs.html#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LDpred2-grid</span></span>
<span id="cb110-2"><a href="polygenic-scores-pgs.html#cb110-2" aria-hidden="true" tabindex="-1"></a>(h2_seq <span class="ot">&lt;-</span> <span class="fu">round</span>(ldsc_h2_est <span class="sc">*</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.7</span>, <span class="dv">1</span>, <span class="fl">1.4</span>), <span class="dv">4</span>))</span></code></pre></div>
<pre><code>#&gt; [1] 0.0158 0.0370 0.0528 0.0740</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="polygenic-scores-pgs.html#cb112-1" aria-hidden="true" tabindex="-1"></a>(p_seq <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fu">seq_log</span>(<span class="fl">1e-5</span>, <span class="dv">1</span>, <span class="at">length.out =</span> <span class="dv">21</span>), <span class="dv">2</span>))</span></code></pre></div>
<pre><code>#&gt;  [1] 1.0e-05 1.8e-05 3.2e-05 5.6e-05 1.0e-04 1.8e-04 3.2e-04 5.6e-04
#&gt;  [9] 1.0e-03 1.8e-03 3.2e-03 5.6e-03 1.0e-02 1.8e-02 3.2e-02 5.6e-02
#&gt; [17] 1.0e-01 1.8e-01 3.2e-01 5.6e-01 1.0e+00</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="polygenic-scores-pgs.html#cb114-1" aria-hidden="true" tabindex="-1"></a>params <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">p =</span> p_seq, <span class="at">h2 =</span> h2_seq, <span class="at">sparse =</span> <span class="fu">c</span>(<span class="cn">FALSE</span>, <span class="cn">TRUE</span>))</span>
<span id="cb114-2"><a href="polygenic-scores-pgs.html#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(params)</span></code></pre></div>
<pre><code>#&gt; [1] 168   3</code></pre>
<p>Here, we will be using this smaller grid instead (for speed in this tutorial):</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="polygenic-scores-pgs.html#cb116-1" aria-hidden="true" tabindex="-1"></a>(params <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">p =</span> <span class="fu">signif</span>(<span class="fu">seq_log</span>(<span class="fl">1e-4</span>, <span class="fl">0.5</span>, <span class="at">length.out =</span> <span class="dv">16</span>), <span class="dv">2</span>),</span>
<span id="cb116-2"><a href="polygenic-scores-pgs.html#cb116-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">h2 =</span> <span class="fu">round</span>(ldsc_h2_est, <span class="dv">4</span>), <span class="at">sparse =</span> <span class="cn">TRUE</span>))</span></code></pre></div>
<pre><code>#&gt;          p     h2 sparse
#&gt; 1  0.00010 0.0528   TRUE
#&gt; 2  0.00018 0.0528   TRUE
#&gt; 3  0.00031 0.0528   TRUE
#&gt; 4  0.00055 0.0528   TRUE
#&gt; 5  0.00097 0.0528   TRUE
#&gt; 6  0.00170 0.0528   TRUE
#&gt; 7  0.00300 0.0528   TRUE
#&gt; 8  0.00530 0.0528   TRUE
#&gt; 9  0.00940 0.0528   TRUE
#&gt; 10 0.01700 0.0528   TRUE
#&gt; 11 0.02900 0.0528   TRUE
#&gt; 12 0.05200 0.0528   TRUE
#&gt; 13 0.09100 0.0528   TRUE
#&gt; 14 0.16000 0.0528   TRUE
#&gt; 15 0.28000 0.0528   TRUE
#&gt; 16 0.50000 0.0528   TRUE</code></pre>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="polygenic-scores-pgs.html#cb118-1" aria-hidden="true" tabindex="-1"></a>beta_grid <span class="ot">&lt;-</span> <span class="fu">snp_ldpred2_grid</span>(corr, df_beta, params, <span class="at">ncores =</span> NCORES)</span>
<span id="cb118-2"><a href="polygenic-scores-pgs.html#cb118-2" aria-hidden="true" tabindex="-1"></a>params<span class="sc">$</span>sparsity <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(beta_grid <span class="sc">==</span> <span class="dv">0</span>)</span></code></pre></div>
<p>Then, we can compute the corresponding PGS for all these models:</p>
<div class="sourceCode" id="cb119"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="polygenic-scores-pgs.html#cb119-1" aria-hidden="true" tabindex="-1"></a>pred_grid <span class="ot">&lt;-</span> <span class="fu">big_prodMat</span>(G, beta_grid, <span class="at">ind.col =</span> df_beta[[<span class="st">&quot;_NUM_ID_&quot;</span>]],</span>
<span id="cb119-2"><a href="polygenic-scores-pgs.html#cb119-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">ncores =</span> NCORES)</span>
<span id="cb119-3"><a href="polygenic-scores-pgs.html#cb119-3" aria-hidden="true" tabindex="-1"></a>params<span class="sc">$</span>score <span class="ot">&lt;-</span> <span class="fu">apply</span>(pred_grid, <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb119-4"><a href="polygenic-scores-pgs.html#cb119-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">is.na</span>(x))) <span class="fu">return</span>(<span class="cn">NA</span>)</span>
<span id="cb119-5"><a href="polygenic-scores-pgs.html#cb119-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>(<span class="fu">glm</span>(CAD <span class="sc">~</span> x <span class="sc">+</span> sex <span class="sc">+</span> age, <span class="at">data =</span> obj.bigsnp<span class="sc">$</span>fam, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>))<span class="sc">$</span>coef[<span class="st">&quot;x&quot;</span>, <span class="dv">3</span>]</span>
<span id="cb119-6"><a href="polygenic-scores-pgs.html#cb119-6" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<p>Note that missing values represent models that diverged substantially.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="polygenic-scores-pgs.html#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(params, <span class="fu">aes</span>(<span class="at">x =</span> p, <span class="at">y =</span> score, <span class="at">color =</span> <span class="fu">as.factor</span>(h2))) <span class="sc">+</span></span>
<span id="cb120-2"><a href="polygenic-scores-pgs.html#cb120-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>() <span class="sc">+</span></span>
<span id="cb120-3"><a href="polygenic-scores-pgs.html#cb120-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb120-4"><a href="polygenic-scores-pgs.html#cb120-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb120-5"><a href="polygenic-scores-pgs.html#cb120-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">breaks =</span> <span class="dv">10</span><span class="sc">^</span>(<span class="sc">-</span><span class="dv">5</span><span class="sc">:</span><span class="dv">0</span>), <span class="at">minor_breaks =</span> params<span class="sc">$</span>p) <span class="sc">+</span></span>
<span id="cb120-6"><a href="polygenic-scores-pgs.html#cb120-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sparse, <span class="at">labeller =</span> label_both) <span class="sc">+</span></span>
<span id="cb120-7"><a href="polygenic-scores-pgs.html#cb120-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;GLM Z-Score&quot;</span>, <span class="at">color =</span> <span class="st">&quot;h2&quot;</span>) <span class="sc">+</span></span>
<span id="cb120-8"><a href="polygenic-scores-pgs.html#cb120-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;top&quot;</span>, <span class="at">panel.spacing =</span> <span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">&quot;lines&quot;</span>))</span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-19-1.png" width="70%" style="display: block; margin: auto;" />
Then you can use the best-performing model here.
Note that you should use only individuals from the validation set to compute the <code>$score</code> and then evaluate the best model for the individuals in the test set.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="polygenic-scores-pgs.html#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb121-2"><a href="polygenic-scores-pgs.html#cb121-2" aria-hidden="true" tabindex="-1"></a>best_beta_grid <span class="ot">&lt;-</span> params <span class="sc">%&gt;%</span></span>
<span id="cb121-3"><a href="polygenic-scores-pgs.html#cb121-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">row_number</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb121-4"><a href="polygenic-scores-pgs.html#cb121-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(score)) <span class="sc">%&gt;%</span></span>
<span id="cb121-5"><a href="polygenic-scores-pgs.html#cb121-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb121-6"><a href="polygenic-scores-pgs.html#cb121-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(id) <span class="sc">%&gt;%</span></span>
<span id="cb121-7"><a href="polygenic-scores-pgs.html#cb121-7" aria-hidden="true" tabindex="-1"></a>  beta_grid[, .]</span></code></pre></div>
<p>To run LDpred2-auto, you can use:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="polygenic-scores-pgs.html#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="co"># LDpred2-auto</span></span>
<span id="cb122-2"><a href="polygenic-scores-pgs.html#cb122-2" aria-hidden="true" tabindex="-1"></a>multi_auto <span class="ot">&lt;-</span> <span class="fu">snp_ldpred2_auto</span>(</span>
<span id="cb122-3"><a href="polygenic-scores-pgs.html#cb122-3" aria-hidden="true" tabindex="-1"></a>  corr, df_beta, <span class="at">h2_init =</span> ldsc_h2_est,</span>
<span id="cb122-4"><a href="polygenic-scores-pgs.html#cb122-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">vec_p_init =</span> <span class="fu">seq_log</span>(<span class="fl">1e-4</span>, <span class="fl">0.2</span>, <span class="dv">30</span>),</span>
<span id="cb122-5"><a href="polygenic-scores-pgs.html#cb122-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">burn_in =</span> <span class="dv">100</span>, <span class="at">num_iter =</span> <span class="dv">100</span>,         <span class="co"># TO REMOVE, for speed here</span></span>
<span id="cb122-6"><a href="polygenic-scores-pgs.html#cb122-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">allow_jump_sign =</span> <span class="cn">FALSE</span>,</span>
<span id="cb122-7"><a href="polygenic-scores-pgs.html#cb122-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">shrink_corr =</span> <span class="fl">0.95</span>,</span>
<span id="cb122-8"><a href="polygenic-scores-pgs.html#cb122-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncores =</span> NCORES)</span></code></pre></div>
<p>Perform some quality control on the chains:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="polygenic-scores-pgs.html#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="co"># `range` should be between 0 and 2</span></span>
<span id="cb123-2"><a href="polygenic-scores-pgs.html#cb123-2" aria-hidden="true" tabindex="-1"></a>(range <span class="ot">&lt;-</span> <span class="fu">sapply</span>(multi_auto, <span class="cf">function</span>(auto) <span class="fu">diff</span>(<span class="fu">range</span>(auto<span class="sc">$</span>corr_est))))</span></code></pre></div>
<pre><code>#&gt;  [1] 0.05495982 0.05372239 0.05382592 0.05373679 0.05246134 0.05391799
#&gt;  [7] 0.05335826 0.05374448 0.05408486 0.05228111 0.05518666 0.05399642
#&gt; [13] 0.05319311 0.05452853 0.05345489 0.05340946 0.05523593 0.05341814
#&gt; [19] 0.05432542 0.05519659 0.05508742 0.05415670 0.05174080 0.05220991
#&gt; [25] 0.05260395 0.05317630 0.04996561 0.05107687 0.05268744 0.05282274</code></pre>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="polygenic-scores-pgs.html#cb125-1" aria-hidden="true" tabindex="-1"></a>(keep <span class="ot">&lt;-</span> (range <span class="sc">&gt;</span> (<span class="fl">0.95</span> <span class="sc">*</span> <span class="fu">quantile</span>(range, <span class="fl">0.95</span>))))</span></code></pre></div>
<pre><code>#&gt;  [1]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
#&gt; [12]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
#&gt; [23] FALSE FALSE  TRUE  TRUE FALSE FALSE  TRUE  TRUE</code></pre>
<p>To get the final effects / predictions, <strong>you should only use chains that pass this filtering</strong>:</p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="polygenic-scores-pgs.html#cb127-1" aria-hidden="true" tabindex="-1"></a>final_beta_auto <span class="ot">&lt;-</span> </span>
<span id="cb127-2"><a href="polygenic-scores-pgs.html#cb127-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowMeans</span>(<span class="fu">sapply</span>(multi_auto[keep], <span class="cf">function</span>(auto) auto<span class="sc">$</span>beta_est))</span></code></pre></div>
<p>We can finally test the final prediction</p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="polygenic-scores-pgs.html#cb128-1" aria-hidden="true" tabindex="-1"></a>final_pred_auto <span class="ot">&lt;-</span> <span class="fu">big_prodVec</span>(G, final_beta_auto,</span>
<span id="cb128-2"><a href="polygenic-scores-pgs.html#cb128-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">ind.col =</span> df_beta[[<span class="st">&quot;_NUM_ID_&quot;</span>]],</span>
<span id="cb128-3"><a href="polygenic-scores-pgs.html#cb128-3" aria-hidden="true" tabindex="-1"></a>                               <span class="at">ncores =</span> NCORES)</span>
<span id="cb128-4"><a href="polygenic-scores-pgs.html#cb128-4" aria-hidden="true" tabindex="-1"></a><span class="fu">AUCBoot</span>(final_pred_auto, obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>CAD)</span></code></pre></div>
<pre><code>#&gt;       Mean       2.5%      97.5%         Sd 
#&gt; 0.56684174 0.53486242 0.59916384 0.01640011</code></pre>
</div>
<div id="lassosum2-grid-of-models" class="section level3 hasAnchor" number="7.1.3">
<h3><span class="header-section-number">7.1.3</span> lassosum2: grid of models<a href="polygenic-scores-pgs.html#lassosum2-grid-of-models" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>lassosum2 is a re-implementation of the lassosum model that now uses the exact same input parameters as LDpred2 (<code>corr</code> and <code>df_beta</code>). It can therefore be run next to LDpred2 and the best model can be chosen using the validation set.
Note that parameter âsâ from lassosum has been replaced by a new parameter âdeltaâ in lassosum2, in order to better reflect that the lassosum model also uses L2-regularization (therefore, elastic-net regularization).</p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb130-1"><a href="polygenic-scores-pgs.html#cb130-1" aria-hidden="true" tabindex="-1"></a>beta_lassosum2 <span class="ot">&lt;-</span> <span class="fu">snp_lassosum2</span>(</span>
<span id="cb130-2"><a href="polygenic-scores-pgs.html#cb130-2" aria-hidden="true" tabindex="-1"></a>  corr, df_beta, <span class="at">ncores =</span> NCORES,</span>
<span id="cb130-3"><a href="polygenic-scores-pgs.html#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">nlambda =</span> <span class="dv">10</span>, <span class="at">maxiter =</span> <span class="dv">50</span>)      <span class="co"># TO REMOVE, for speed here</span></span></code></pre></div>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="polygenic-scores-pgs.html#cb131-1" aria-hidden="true" tabindex="-1"></a>params2 <span class="ot">&lt;-</span> <span class="fu">attr</span>(beta_lassosum2, <span class="st">&quot;grid_param&quot;</span>)</span>
<span id="cb131-2"><a href="polygenic-scores-pgs.html#cb131-2" aria-hidden="true" tabindex="-1"></a>pred_grid2 <span class="ot">&lt;-</span> <span class="fu">big_prodMat</span>(G, beta_lassosum2, <span class="at">ind.col =</span> df_beta[[<span class="st">&quot;_NUM_ID_&quot;</span>]],</span>
<span id="cb131-3"><a href="polygenic-scores-pgs.html#cb131-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ncores =</span> NCORES)</span>
<span id="cb131-4"><a href="polygenic-scores-pgs.html#cb131-4" aria-hidden="true" tabindex="-1"></a>params2<span class="sc">$</span>score <span class="ot">&lt;-</span> <span class="fu">apply</span>(pred_grid2, <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb131-5"><a href="polygenic-scores-pgs.html#cb131-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">all</span>(<span class="fu">is.na</span>(x))) <span class="fu">return</span>(<span class="cn">NA</span>)</span>
<span id="cb131-6"><a href="polygenic-scores-pgs.html#cb131-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>(<span class="fu">glm</span>(CAD <span class="sc">~</span> x <span class="sc">+</span> sex <span class="sc">+</span> age, <span class="at">data =</span> obj.bigsnp<span class="sc">$</span>fam, <span class="at">family =</span> <span class="st">&quot;binomial&quot;</span>))<span class="sc">$</span>coef[<span class="st">&quot;x&quot;</span>, <span class="dv">3</span>]</span>
<span id="cb131-7"><a href="polygenic-scores-pgs.html#cb131-7" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="polygenic-scores-pgs.html#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(params2, <span class="fu">aes</span>(<span class="at">x =</span> lambda, <span class="at">y =</span> score, <span class="at">color =</span> <span class="fu">as.factor</span>(delta))) <span class="sc">+</span></span>
<span id="cb132-2"><a href="polygenic-scores-pgs.html#cb132-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>() <span class="sc">+</span></span>
<span id="cb132-3"><a href="polygenic-scores-pgs.html#cb132-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb132-4"><a href="polygenic-scores-pgs.html#cb132-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb132-5"><a href="polygenic-scores-pgs.html#cb132-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">breaks =</span> <span class="dv">10</span><span class="sc">^</span>(<span class="sc">-</span><span class="dv">5</span><span class="sc">:</span><span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb132-6"><a href="polygenic-scores-pgs.html#cb132-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">&quot;GLM Z-Score&quot;</span>, <span class="at">color =</span> <span class="st">&quot;delta&quot;</span>)</span></code></pre></div>
<p><img src="pgs_files/figure-html/unnamed-chunk-27-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="polygenic-scores-pgs.html#cb133-1" aria-hidden="true" tabindex="-1"></a>best_grid_lassosum2 <span class="ot">&lt;-</span> params2 <span class="sc">%&gt;%</span></span>
<span id="cb133-2"><a href="polygenic-scores-pgs.html#cb133-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">id =</span> <span class="fu">row_number</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb133-3"><a href="polygenic-scores-pgs.html#cb133-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(score)) <span class="sc">%&gt;%</span></span>
<span id="cb133-4"><a href="polygenic-scores-pgs.html#cb133-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb133-5"><a href="polygenic-scores-pgs.html#cb133-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(id) <span class="sc">%&gt;%</span> </span>
<span id="cb133-6"><a href="polygenic-scores-pgs.html#cb133-6" aria-hidden="true" tabindex="-1"></a>  beta_lassosum2[, .]</span></code></pre></div>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="polygenic-scores-pgs.html#cb134-1" aria-hidden="true" tabindex="-1"></a>best_grid_overall <span class="ot">&lt;-</span> <span class="st">`</span><span class="at">if</span><span class="st">`</span>(<span class="fu">max</span>(params2<span class="sc">$</span>score) <span class="sc">&gt;</span> <span class="fu">max</span>(params<span class="sc">$</span>score),</span>
<span id="cb134-2"><a href="polygenic-scores-pgs.html#cb134-2" aria-hidden="true" tabindex="-1"></a>                          best_grid_lassosum2,</span>
<span id="cb134-3"><a href="polygenic-scores-pgs.html#cb134-3" aria-hidden="true" tabindex="-1"></a>                          best_beta_grid)</span></code></pre></div>

</div>
</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bycroft2018uk" class="csl-entry">
Bycroft, C., Freeman, C., Petkova, D., Band, G., Elliott, L.T., Sharp, K., et al.others. (2018). <a href="https://doi.org/10.1038/s41586-018-0579-z">The <span>UK Biobank</span> resource with deep phenotyping and genomic data</a>. <em>Nature</em>, <em>562</em>, 203â209.
</div>
<div id="ref-prive2022inferring" class="csl-entry">
PrivÃ©, F., AlbiÃ±ana, C., Pasaniuc, B., &amp; VilhjÃ¡lmsson, B.J. (2022). Inferring disease architecture and predictive ability with LDpred2-auto. <em>bioRxiv</em>. Retrieved from <a href="https://doi.org/10.1101/2022.10.10.511629">https://doi.org/10.1101/2022.10.10.511629</a>
</div>
<div id="ref-prive2021identifying" class="csl-entry">
PrivÃ©, F., Arbel, J., Aschard, H., &amp; VilhjÃ¡lmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136">Identifying and correcting for misspecifications in <span>GWAS</span> summary statistics and polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>, <em>3</em>, 100136.
</div>
<div id="ref-prive2020ldpred2" class="csl-entry">
PrivÃ©, F., Arbel, J., &amp; VilhjÃ¡lmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa1029"><span class="nocase"><span>LDpred2</span>: better, faster, stronger</span></a>. <em>Bioinformatics</em>, <em>36</em>, 5424â5431.
</div>
<div id="ref-prive2019efficient" class="csl-entry">
PrivÃ©, F., Aschard, H., &amp; Blum, M.G. (2019). <a href="https://doi.org/10.1534/genetics.119.302019">Efficient implementation of penalized regression for genetic risk prediction</a>. <em>Genetics</em>, <em>212</em>, 65â74.
</div>
<div id="ref-prive2019making" class="csl-entry">
PrivÃ©, F., VilhjÃ¡lmsson, B.J., Aschard, H., &amp; Blum, M.G.B. (2019). <a href="https://doi.org/10.1016/j.ajhg.2019.11.001">Making the most of clumping and thresholding for polygenic scores</a>. <em>The American Journal of Human Genetics</em>, <em>105</em>, 1213â1221.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="genome-wide-association-study-gwas.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="ipsych-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
