<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Preprocessing | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Preprocessing | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://privefl.github.io/bigsnpr-extdoc/" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Preprocessing | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />

<meta name="author" content="Florian PrivÃ©" />


<meta name="date" content="2021-03-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="inputs-and-formats.html"/>
<link rel="next" href="population-structure.html"/>
<script src="libs/header-attrs-2.7.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/font-awesome-5.3.1/css/fontawesome-all.min.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.4</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-a-fbm-or-bigsnp-object"><i class="fa fa-check"></i><b>2.3</b> Getting a FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#conversion-and-quality-control-of-plink-files"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Exercise: preprocessing</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.2</b> Exercise: population structure</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Exercise: GWAS</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html"><i class="fa fa-check"></i><b>6</b> Polygenic scores (PGS)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#exercises-ldpred2-and-lassosum2"><i class="fa fa-check"></i><b>6.1</b> Exercises: LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#preparing-the-data"><i class="fa fa-check"></i><b>6.1.1</b> Preparing the data</a></li>
<li class="chapter" data-level="6.1.2" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#ldpred2"><i class="fa fa-check"></i><b>6.1.2</b> LDpred2</a></li>
<li class="chapter" data-level="6.1.3" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>6.1.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="preprocessing" class="section level1" number="3">
<h1><span class="header-section-number">Chapter 3</span> Preprocessing</h1>
<p>In this section, I am including conversion, quality control and imputation.</p>
<div id="conversion-and-quality-control-of-plink-files" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Conversion and quality control of PLINK files</h2>
<p>PLINK is very efficient for conversion and quality control of multiple formats, so I have decided just using it.</p>
<p>In {bigsnpr}, I provide some wrappers to PLINK for ease of use:</p>
<ul>
<li><p><code>download_plink()</code> and <code>download_plink2()</code> for downloading the latest stable versions of PLINK 1.9 and 2.0.</p></li>
<li><p><code>snp_plinkQC()</code> for quality control (QC) and conversion to bed/bim/fam.</p></li>
<li><p><code>snp_plinkKINGQC()</code> for QC on relatedness based on KING-robust kinship estimator. Using <code>make.bed = FALSE</code> allows for computing related pairs only, i.e.Â reporting a data frame without producing new bed/bim/fam files.</p></li>
<li><p><code>snp_plinkIBDQC()</code> for QC based on identity-by-descent (IBD) computed by PLINK using its method-of-moments.</p></li>
<li><p><code>snp_plinkRmSamples()</code> for producing new PLINK files after having removed some individuals.</p></li>
<li><p>For any other PLINK function, I recommend calling PLINK directly from <i class="fab  fa-r-project "></i> thanks to system calls and package {glue}, e.g.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="preprocessing.html#cb1-1" aria-hidden="true" tabindex="-1"></a>plink <span class="ot">&lt;-</span> <span class="fu">download_plink</span>(<span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb1-2"><a href="preprocessing.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(glue<span class="sc">::</span><span class="fu">glue</span>(</span>
<span id="cb1-3"><a href="preprocessing.html#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;{plink} --version&quot;</span></span>
<span id="cb1-4"><a href="preprocessing.html#cb1-4" aria-hidden="true" tabindex="-1"></a>))</span></code></pre></div>
<pre><code>#&gt; PLINK v1.90b6.21 64-bit (19 Oct 2020)</code></pre></li>
</ul>
</div>
<div id="imputation" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Imputation</h2>
<p>Note that most functions from {bigstatsr} and {bigsnpr} donât handle missing values.</p>
<p>Simple imputation (e.g.Â by the mean) of a âdoubleâ FBM can be performed by blocks using e.g.Â as explained in <a href="https://privefl.github.io/bigstatsr/articles/big-apply.html">this vignette</a>.</p>
<p>In {bigsnpr}, to perform simple imputation of genotyped data, you can use <code>snp_fastImputeSimple()</code>. You can also use <code>snp_fastImpute()</code> that uses XGBoost models to impute genotyped data (slower but still fast enough, described in <span class="citation"><a href="references.html#ref-prive2017efficient" role="doc-biblioref">PrivÃ© et al.</a> (<a href="references.html#ref-prive2017efficient" role="doc-biblioref">2018</a>)</span>).</p>
</div>
<div id="exo-preprocessing" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Exercise: preprocessing</h2>
<p>For the exercises, we will use the data provided in <span class="citation"><a href="references.html#ref-reed2015guide" role="doc-biblioref">Reed et al.</a> (<a href="references.html#ref-reed2015guide" role="doc-biblioref">2015</a>)</span>.</p>
<p>This can be downloaded using</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="preprocessing.html#cb3-1" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">&quot;https://www.mtholyoke.edu/courses/afoulkes/Data/statsTeachR/&quot;</span></span>
<span id="cb3-2"><a href="preprocessing.html#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(<span class="fu">paste0</span>(url, <span class="st">&quot;GWAS_data&quot;</span>, <span class="fu">c</span>(<span class="st">&quot;.bed&quot;</span>, <span class="st">&quot;.bim&quot;</span>, <span class="st">&quot;.fam&quot;</span>)),</span>
<span id="cb3-3"><a href="preprocessing.html#cb3-3" aria-hidden="true" tabindex="-1"></a>       runonce<span class="sc">::</span>download_file, <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>)</span></code></pre></div>
<pre><code>#&gt; https://www.mtholyoke.edu/courses/afoulkes/Data/statsTeachR/GWAS_data.bed 
#&gt;                                                  &quot;tmp-data/GWAS_data.bed&quot; 
#&gt; https://www.mtholyoke.edu/courses/afoulkes/Data/statsTeachR/GWAS_data.bim 
#&gt;                                                  &quot;tmp-data/GWAS_data.bim&quot; 
#&gt; https://www.mtholyoke.edu/courses/afoulkes/Data/statsTeachR/GWAS_data.fam 
#&gt;                                                  &quot;tmp-data/GWAS_data.fam&quot;</code></pre>
<p>For some reason, this data is not ordered by chromosome and position; we can use PLINK to get an ordered version of this using</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="preprocessing.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb5-2"><a href="preprocessing.html#cb5-2" aria-hidden="true" tabindex="-1"></a>plink <span class="ot">&lt;-</span> <span class="fu">download_plink</span>(<span class="st">&quot;tmp-data&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="preprocessing.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(glue<span class="sc">::</span><span class="fu">glue</span>(</span>
<span id="cb6-2"><a href="preprocessing.html#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot;{plink} --bfile tmp-data/GWAS_data&quot;</span>,</span>
<span id="cb6-3"><a href="preprocessing.html#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">&quot; --make-bed --out tmp-data/GWAS_data_sorted&quot;</span></span>
<span id="cb6-4"><a href="preprocessing.html#cb6-4" aria-hidden="true" tabindex="-1"></a>))</span></code></pre></div>
<pre><code>#&gt; PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/
#&gt; (C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
#&gt; Logging to tmp-data/GWAS_data_sorted.log.
#&gt; Options in effect:
#&gt;   --bfile tmp-data/GWAS_data
#&gt;   --make-bed
#&gt;   --out tmp-data/GWAS_data_sorted
#&gt; 
#&gt; 32574 MB RAM detected; reserving 16287 MB for main workspace.
#&gt; 500000 variants loaded from .bim file.
#&gt; 1401 people (937 males, 464 females) loaded from .fam.
#&gt; 933 phenotype values loaded from .fam.
#&gt; Using 1 thread (no multithreaded calculations invoked).
#&gt; Before main variant filters, 1401 founders and 0 nonfounders present.
#&gt; Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99% done.
#&gt; Total genotyping rate is 0.977593.
#&gt; 500000 variants and 1401 people pass filters and QC.
#&gt; Among remaining phenotypes, 463 are cases and 470 are controls.  (468
#&gt; phenotypes are missing.)
#&gt; --make-bed to tmp-data/GWAS_data_sorted.bed + tmp-data/GWAS_data_sorted.bim +
#&gt; tmp-data/GWAS_data_sorted.fam ... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.</code></pre>
<p>As you can see from PLINK output, this data contains 1401 individuals and 500,000 variants, with a few missing values.</p>
<p>We can then perform some quality control using</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="preprocessing.html#cb8-1" aria-hidden="true" tabindex="-1"></a>bedfile2 <span class="ot">&lt;-</span> <span class="fu">snp_plinkQC</span>(plink, <span class="st">&quot;tmp-data/GWAS_data_sorted&quot;</span>)</span></code></pre></div>
<pre><code>#&gt; PLINK v1.90b6.21 64-bit (19 Oct 2020)          www.cog-genomics.org/plink/1.9/
#&gt; (C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3
#&gt; Logging to tmp-data/GWAS_data_sorted_QC.log.
#&gt; Options in effect:
#&gt;   --bfile tmp-data/GWAS_data_sorted
#&gt;   --geno 0.1
#&gt;   --hwe 1e-50
#&gt;   --maf 0.01
#&gt;   --make-bed
#&gt;   --mind 0.1
#&gt;   --out tmp-data/GWAS_data_sorted_QC
#&gt; 
#&gt; 32574 MB RAM detected; reserving 16287 MB for main workspace.
#&gt; 500000 variants loaded from .bim file.
#&gt; 1401 people (937 males, 464 females) loaded from .fam.
#&gt; 933 phenotype values loaded from .fam.
#&gt; 0 people removed due to missing genotype data (--mind).
#&gt; Using 1 thread (no multithreaded calculations invoked).
#&gt; Before main variant filters, 1401 founders and 0 nonfounders present.
#&gt; Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99% done.
#&gt; Total genotyping rate is 0.977593.
#&gt; 27390 variants removed due to missing genotype data (--geno).
#&gt; --hwe: 91 variants removed due to Hardy-Weinberg exact test.
#&gt; 67856 variants removed due to minor allele threshold(s)
#&gt; (--maf/--max-maf/--mac/--max-mac).
#&gt; 404663 variants and 1401 people pass filters and QC.
#&gt; Among remaining phenotypes, 463 are cases and 470 are controls.  (468
#&gt; phenotypes are missing.)
#&gt; --make-bed to tmp-data/GWAS_data_sorted_QC.bed +
#&gt; tmp-data/GWAS_data_sorted_QC.bim + tmp-data/GWAS_data_sorted_QC.fam ... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.</code></pre>
<p>404,663 variants are remaining after this quality control; we can then read this data into an R object called <code>bigSNP</code> using</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="preprocessing.html#cb10-1" aria-hidden="true" tabindex="-1"></a>(rds <span class="ot">&lt;-</span> <span class="fu">snp_readBed2</span>(bedfile2, <span class="at">ncores =</span> <span class="fu">nb_cores</span>()))</span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\Desktop\\bigsnpr-extdoc\\tmp-data\\GWAS_data_sorted_QC.rds&quot;</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="preprocessing.html#cb12-1" aria-hidden="true" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(rds)</span>
<span id="cb12-2"><a href="preprocessing.html#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(obj.bigsnp, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code></pre></div>
<pre><code>#&gt; List of 3
#&gt;  $ genotypes:Reference class &#39;FBM.code256&#39; [package &quot;bigstatsr&quot;] with 16 fields
#&gt;   ..and 26 methods, of which 12 are  possibly relevant:
#&gt;   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
#&gt;   ..  check_write_permissions, copy#envRefClass, initialize,
#&gt;   ..  initialize#FBM, save, show#envRefClass, show#FBM
#&gt;  $ fam      :&#39;data.frame&#39;:   1401 obs. of  6 variables:
#&gt;   ..$ family.ID  : int [1:1401] 10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...
#&gt;   ..$ sample.ID  : int [1:1401] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ paternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:1401] 1 2 1 1 1 1 1 2 1 2 ...
#&gt;   ..$ affection  : int [1:1401] 1 1 2 1 2 2 2 1 2 -9 ...
#&gt;  $ map      :&#39;data.frame&#39;:   404663 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:404663] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:404663] &quot;rs12565286&quot; &quot;rs3094315&quot; &quot;rs2980319&quot; &quot;rs2980300&quot; ...
#&gt;   ..$ genetic.dist: int [1:404663] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:404663] 721290 752566 777122 785989 798959 947034 949608 1018704 1041700 1129672 ...
#&gt;   ..$ allele1     : chr [1:404663] &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;   ..$ allele2     : chr [1:404663] &quot;C&quot; &quot;T&quot; &quot;T&quot; &quot;G&quot; ...
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;bigSNP&quot;</code></pre>
<p>We can store some extra information on the individuals (e.g.Â some phenotypes):</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="preprocessing.html#cb14-1" aria-hidden="true" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb14-2"><a href="preprocessing.html#cb14-2" aria-hidden="true" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(<span class="fu">paste0</span>(url, <span class="st">&quot;GWAS_clinical.csv&quot;</span>), <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>))</span>
<span id="cb14-3"><a href="preprocessing.html#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the same order as for the genotypes</span></span>
<span id="cb14-4"><a href="preprocessing.html#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># (to match over multiple columns, use `vctrs::vec_match()`)</span></span>
<span id="cb14-5"><a href="preprocessing.html#cb14-5" aria-hidden="true" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> clinical[<span class="fu">match</span>(obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>family.ID, clinical<span class="sc">$</span>FamID), ]</span>
<span id="cb14-6"><a href="preprocessing.html#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Some verif</span></span>
<span id="cb14-7"><a href="preprocessing.html#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>sex, pheno<span class="sc">$</span>sex))</span>
<span id="cb14-8"><a href="preprocessing.html#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Update the $fam component</span></span>
<span id="cb14-9"><a href="preprocessing.html#cb14-9" aria-hidden="true" tabindex="-1"></a>obj.bigsnp<span class="sc">$</span>fam <span class="ot">&lt;-</span> <span class="fu">cbind</span>(obj.bigsnp<span class="sc">$</span>fam, pheno[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)])</span></code></pre></div>
<p>Recall that this data contains some missing values; you can get some counts per variant using</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="preprocessing.html#cb15-1" aria-hidden="true" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb15-2"><a href="preprocessing.html#cb15-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">big_counts</span>(G)</span>
<span id="cb15-3"><a href="preprocessing.html#cb15-3" aria-hidden="true" tabindex="-1"></a>counts[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;   17   15    0   18    1   78   64   69</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="preprocessing.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(nbNA <span class="ot">&lt;-</span> counts[<span class="dv">4</span>, ])</span></code></pre></div>
<p><img src="preprocessing_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can e.g.Â perform a quick imputation by the mean using</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="preprocessing.html#cb18-1" aria-hidden="true" tabindex="-1"></a>G2 <span class="ot">&lt;-</span> <span class="fu">snp_fastImputeSimple</span>(G, <span class="at">method =</span> <span class="st">&quot;mean2&quot;</span>, <span class="at">ncores =</span> <span class="fu">nb_cores</span>())</span>
<span id="cb18-2"><a href="preprocessing.html#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">big_counts</span>(G2, <span class="at">ind.col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;    0    0    0    0    0    0    0    0
#&gt; 0.01    0    0    0    0    0    0    0    0
#&gt; 0.02    0    0    0    0    0    0    0    0
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 196 rows ]</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="preprocessing.html#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">big_counts</span>(G, <span class="at">ind.col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;   17   15    0   18    1   78   64   69</code></pre>
<p><code>G</code> still has missing values, but <code>G2</code> does not. Note that both are using the same data, the difference is that they using a different code to decode the underlying data:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="preprocessing.html#cb22-1" aria-hidden="true" tabindex="-1"></a>G<span class="sc">$</span>code256</span></code></pre></div>
<pre><code>#&gt;  [1]  0  1  2 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
#&gt; [24] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
#&gt; [47] NA NA NA NA
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 206 entries ]</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="preprocessing.html#cb24-1" aria-hidden="true" tabindex="-1"></a>G2<span class="sc">$</span>code256</span></code></pre></div>
<pre><code>#&gt;  [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04 0.05 0.06
#&gt; [15] 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18 0.19 0.20
#&gt; [29] 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31 0.32 0.33 0.34
#&gt; [43] 0.35 0.36 0.37 0.38 0.39 0.40 0.41 0.42
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 206 entries ]</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="preprocessing.html#cb26-1" aria-hidden="true" tabindex="-1"></a>obj.bigsnp<span class="sc">$</span>genotypes <span class="ot">&lt;-</span> G2</span></code></pre></div>
<p>To always use <code>G2</code> (with the new <code>code256</code>) and the extended <code>obj.bigsnp$fam</code>, you need to re-save <code>obj.bigsnp</code> using</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="preprocessing.html#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">snp_save</span>(obj.bigsnp)</span></code></pre></div>
<p>You can re-attach this data in another R session later using <code>snp_attach("tmp-data/GWAS_data_sorted_QC.rds")</code>.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="inputs-and-formats.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="population-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
