<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | bigsnpr &amp; bigstatsr</title>
  <meta name="description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="References | bigsnpr &amp; bigstatsr" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://privefl.github.io/bigsnpr-extdoc/" />
  <meta property="og:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />
  <meta property="og:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="github-repo" content="privefl/bigsnpr-extdoc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | bigsnpr &amp; bigstatsr" />
  <meta name="twitter:site" content="@privefl" />
  <meta name="twitter:description" content="Extended documentation for bigsnpr &amp; bigstatsr." />
  <meta name="twitter:image" content="https://privefl.github.io/bigsnpr-extdoc/images/bigstatsr.png" />

<meta name="author" content="Florian Privé" />


<meta name="date" content="2021-03-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="polygenic-scores-pgs.html"/>

<script src="libs/header-attrs-2.7.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/font-awesome-5.3.1/css/fontawesome-all.min.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-66N68ENQ12"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-66N68ENQ12');
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">bigsnpr & bigstatsr</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#main-motivation-for-developing-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>1.1</b> Main motivation for developing {bigstatsr} and {bigsnpr}</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#features"><i class="fa fa-check"></i><b>1.2</b> Features</a></li>
<li class="chapter" data-level="1.3" data-path="introduction.html"><a href="introduction.html#installation"><i class="fa fa-check"></i><b>1.3</b> Installation</a></li>
<li class="chapter" data-level="1.4" data-path="introduction.html"><a href="introduction.html#correct-spellings"><i class="fa fa-check"></i><b>1.4</b> Correct spellings</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigstatsr"><i class="fa fa-check"></i><b>2.1</b> In {bigstatsr}</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#in-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> In {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-a-fbm-or-bigsnp-object"><i class="fa fa-check"></i><b>2.3</b> Getting a FBM or bigSNP object</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#conversion-and-quality-control-of-plink-files"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Exercise: preprocessing</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.2</b> Exercise: population structure</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Exercise: GWAS</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html"><i class="fa fa-check"></i><b>6</b> Polygenic scores (PGS)</a>
<ul>
<li class="chapter" data-level="6.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#exercises-ldpred2-and-lassosum2"><i class="fa fa-check"></i><b>6.1</b> Exercises: LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#preparing-the-data"><i class="fa fa-check"></i><b>6.1.1</b> Preparing the data</a></li>
<li class="chapter" data-level="6.1.2" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#ldpred2"><i class="fa fa-check"></i><b>6.1.2</b> LDpred2</a></li>
<li class="chapter" data-level="6.1.3" data-path="polygenic-scores-pgs.html"><a href="polygenic-scores-pgs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>6.1.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">bigsnpr &amp; bigstatsr</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered">
<h1>References</h1>

<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-aschard2015adjusting" class="csl-entry">
Aschard, Hugues, Bjarni J Vilhjálmsson, Amit D Joshi, Alkes L Price, and Peter Kraft. 2015. <span>“Adjusting for Heritable Covariates Can Bias Effect Estimates in Genome-Wide Association Studies.”</span> <em>The American Journal of Human Genetics</em> 96 (2): 329–39.
</div>
<div id="ref-buniello2019nhgri" class="csl-entry">
Buniello, Annalisa, Jacqueline A L MacArthur, Maria Cerezo, Laura W Harris, James Hayhurst, Cinzia Malangone, Aoife McMahon, et al. 2019. <span>“The <span>NHGRI-EBI GWAS Catalog</span> of Published Genome-Wide Association Studies, Targeted Arrays and Summary Statistics 2019.”</span> <em>Nucleic Acids Research</em> 47 (D1): D1005–12.
</div>
<div id="ref-bycroft2018uk" class="csl-entry">
Bycroft, Clare, Colin Freeman, Desislava Petkova, Gavin Band, Lloyd T Elliott, Kevin Sharp, Allan Motyer, et al. 2018. <span>“The <span>UK Biobank</span> Resource with Deep Phenotyping and Genomic Data.”</span> <em>Nature</em> 562 (7726): 203–9.
</div>
<div id="ref-nikpay2015comprehensive" class="csl-entry">
Nikpay, Majid, Anuj Goel, Hong-Hee Won, Leanne M Hall, Christina Willenborg, Stavroula Kanoni, Danish Saleheen, et al. 2015. <span>“A Comprehensive 1000 Genomes–Based Genome-Wide Association Meta-Analysis of Coronary Artery Disease.”</span> <em>Nature Genetics</em> 47 (10): 1121.
</div>
<div id="ref-prive2021optimal" class="csl-entry">
Privé, Florian. 2021. <span>“Optimal Linkage Disequilibrium Splitting.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-prive2020ldpred2" class="csl-entry">
Privé, Florian, Julyan Arbel, and Bjarni J Vilhjálmsson. 2020. <span>“<span class="nocase"><span>LDpred2</span>: better, faster, stronger</span>.”</span> <em>Bioinformatics</em>, December. <a href="https://doi.org/10.1093/bioinformatics/btaa1029">https://doi.org/10.1093/bioinformatics/btaa1029</a>.
</div>
<div id="ref-prive2019efficient" class="csl-entry">
Privé, Florian, Hugues Aschard, and Michael GB Blum. 2019. <span>“Efficient Implementation of Penalized Regression for Genetic Risk Prediction.”</span> <em>Genetics</em> 212 (1): 65–74.
</div>
<div id="ref-prive2021high" class="csl-entry">
Privé, Florian, Hugues Aschard, Shai Carmi, Lasse Folkersen, Clive Hoggart, Paul F O’Reilly, and Bjarni J Vilhjálmsson. 2021. <span>“High-Resolution Portability of 245 Polygenic Scores When Derived and Applied in the Same Cohort.”</span> <em>medRxiv</em>.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
Privé, Florian, Hugues Aschard, Andrey Ziyatdinov, and Michael G B Blum. 2018. <span>“Efficient Analysis of Large-Scale Genome-Wide Data with Two <span>R</span> Packages: Bigstatsr and Bigsnpr.”</span> <em>Bioinformatics</em> 34 (16): 2781–87. <a href="https://doi.org/10.1093/bioinformatics/bty185">https://doi.org/10.1093/bioinformatics/bty185</a>.
</div>
<div id="ref-prive2020efficient" class="csl-entry">
Privé, Florian, Keurcien Luu, Michael G B Blum, John J McGrath, and Bjarni J Vilhjálmsson. 2020. <span>“<span class="nocase">Efficient toolkit implementing best practices for principal component analysis of population genetic data</span>.”</span> <em>Bioinformatics</em>, May. <a href="https://doi.org/10.1093/bioinformatics/btaa520">https://doi.org/10.1093/bioinformatics/btaa520</a>.
</div>
<div id="ref-prive2020performing" class="csl-entry">
Privé, Florian, Keurcien Luu, Bjarni J Vilhjálmsson, and Michael GB Blum. 2020. <span>“Performing Highly Efficient Genome Scans for Local Adaptation with <span>R</span> Package Pcadapt Version 4.”</span> <em>Molecular Biology and Evolution</em> 37 (7): 2153–54.
</div>
<div id="ref-prive2019making" class="csl-entry">
Privé, Florian, Bjarni J Vilhjálmsson, Hugues Aschard, and Michael G. B. Blum. 2019. <span>“Making the Most of Clumping and Thresholding for Polygenic Scores.”</span> <em>The American Journal of Human Genetics</em> 105 (6): 1213–21.
</div>
<div id="ref-reed2015guide" class="csl-entry">
Reed, Eric, Sara Nunez, David Kulp, Jing Qian, Muredach P Reilly, and Andrea S Foulkes. 2015. <span>“A Guide to Genome-Wide Association Analysis and Post-Analytic Interrogation.”</span> <em>Statistics in Medicine</em> 34 (28): 3769–92.
</div>
<div id="ref-visscher201710" class="csl-entry">
Visscher, Peter M, Naomi R Wray, Qian Zhang, Pamela Sklar, Mark I McCarthy, Matthew A Brown, and Jian Yang. 2017. <span>“10 Years of GWAS Discovery: Biology, Function, and Translation.”</span> <em>The American Journal of Human Genetics</em> 101 (1): 5–22.
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="polygenic-scores-pgs.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
